# Overview
This repository contains the code used in the SSR vs. SNP marker comparison project,
one of the initiatives in the GCCO group. We're looking to compare how microsatellite
and RADseq markers differ in their measurements of ex situ representation
of two different oak species, _Quercus boyntonii_ (or QUBO; Boynton oak, in Alabama, in the white oak
section of the genus) and _Quercus acerifolia_ (or QUAC; a red oak, section Lobatae). 
Both of these species are included on the IUCN Red List.

Below we describe the general structure of the repository and it's directories. Generally,
this repo will contain the scripts used to generate data for different steps in our analyses
and any other input files required for those scripts to run, such as parameter files, population
files, and input files (when GH size limits allow). 

### denovoAssembly
This contains the scripts and parameter files used to explore the parameter space of a Stacks 
_de novo_ assembly.

### filtering_QC
This folder contains various scripts used to clean the raw NextRAD data (generated from the
sequencing company SNPsaurus). The fastQC folder contains the output of the `fastqc` program,
which was run on our raw data, our data cleaned with `process_radtags` *without* adapter trimming
("Untrimmed"), and our data cleaned with `process_radtags` *with* adapter trimming. 

The preliminarySubset folder contains the scripts and Stacks files used to analyze a 12 sample subset
of QUBO and QUAC samples together. The purpose of the preliminarySubset analysis was to check for 
basic assumptions of our data: making sure replicates matched, and making sure samples within the same
populations were more similar to each other than samples across populations. There are subfolders for 
the scripts passed to Stacks, STRUCTURE (for analyzing the Stacks output), and R (also for analyzing 
the Stacks output)

Finally, the Barcodes.txt file was used by the `process_radtags` command to sort samples by their species
and collection origin (Garden, G, or Wild, W). The dataCleaning script contains the command `process_radtags`
commands sent to Stacks, as well as other commands used to generate and organize Stacks output.

### referenceAlignment
This folder contains the files used to align "raw" sequences (which have actually been passed through 
`process_radtags`) to a reference genome, as well as the consensus sequences generated by the Stacks
*de novo* assembly. Also within this folder will be the scripts used to process both of these alignments
using Stacks.

## Contact
For questions about this dataset, open an Issue or contact Austin. 

A link to the raw data for this analysis can be found on the NCBI SRA [here](https://submit.ncbi.nlm.nih.gov/subs/sra/SUB10415299/overview)
