# %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
# %%% ANALYZING MISSING DATA IN STRUCTURE %%%
# %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

# This script reads in the output files generated by certain STRUCTURE runs to 
# generate a table of missing data values. The purpose of this script is to 
# determine whether certain samples have consistently high levels of missing data 
# across STRUCTURE runs (and how they may be impacting analyses)

# library(pophelper)

# %%%% QUAC %%%% ----
# Garden and wild, R80, First SNP per locus ----
QUAC_str_output_wd <- "/RAID1/IMLS_GCCO/Analysis/structure/denovo_finalAssemblies/QUAC/R80/garden_wild/firstSNP/output/"
setwd(QUAC_str_output_wd)
# Read in and reformat STRUCTURE output files ---- 
# K=2
K2_output <- "QUAC_DNFA_R80_NOMAF_1SNP_2Pops_2-1_f"
QUAC.K_2 <- read.table(K2_output, skip = grep("Inferred ancestry of individuals",
                                              x = readLines(K2_output)), 
                       header = TRUE, nrows = 193)
QUAC.K_2 <- QUAC.K_2[,-(3:5)]
QUAC.K_2[,2] <- gsub("\\(", "", QUAC.K_2[,2])
QUAC.K_2[,2] <- gsub("\\)", "", QUAC.K_2[,2])
# K=3
K3_output <- "QUAC_DNFA_R80_NOMAF_1SNP_2Pops_3-1_f"
QUAC.K_3 <- read.table(K3_output, skip = grep("Inferred ancestry of individuals",
                                              x = readLines(K3_output))+1,
                       nrows = 193)

QUAC.K_3 <- QUAC.K_3[,-(3:5)]
QUAC.K_3[,2] <- gsub("\\(", "", QUAC.K_3[,2])
QUAC.K_3[,2] <- gsub("\\)", "", QUAC.K_3[,2])
# K=4
K2_output <- "QUAC_DNFA_R80_NOMAF_1SNP_2Pops_2-1_f"
QUAC.K_2 <- read.table(K2_output, skip = grep("Inferred ancestry of individuals",
                                              x = readLines(K2_output)), 
                       header = TRUE, nrows = 193)
QUAC.K_2 <- QUAC.K_2[,-(3:5)]
QUAC.K_2[,2] <- gsub("\\(", "", QUAC.K_2[,2])
QUAC.K_2[,2] <- gsub("\\)", "", QUAC.K_2[,2])
# K=5
K2_output <- "QUAC_DNFA_R80_NOMAF_1SNP_2Pops_2-1_f"
QUAC.K_2 <- read.table(K2_output, skip = grep("Inferred ancestry of individuals",
                                              x = readLines(K2_output)), 
                       header = TRUE, nrows = 193)
QUAC.K_2 <- QUAC.K_2[,-(3:5)]
QUAC.K_2[,2] <- gsub("\\(", "", QUAC.K_2[,2])
QUAC.K_2[,2] <- gsub("\\)", "", QUAC.K_2[,2])
# K=6
K2_output <- "QUAC_DNFA_R80_NOMAF_1SNP_2Pops_2-1_f"
QUAC.K_2 <- read.table(K2_output, skip = grep("Inferred ancestry of individuals",
                                              x = readLines(K2_output)), 
                       header = TRUE, nrows = 193)
QUAC.K_2 <- QUAC.K_2[,-(3:5)]
QUAC.K_2[,2] <- gsub("\\(", "", QUAC.K_2[,2])
QUAC.K_2[,2] <- gsub("\\)", "", QUAC.K_2[,2])
# K=7
K7_output <- "QUAC_DNFA_R80_NOMAF_1SNP_2Pops_7-1_f"
QUAC.K_7 <- read.table(K7_output, skip = grep("Inferred ancestry of individuals",
                                              x = readLines(K7_output)), 
                       header = TRUE, nrows = 193)
QUAC.K_7 <- QUAC.K_7[,-(3:5)]
QUAC.K_7[,2] <- gsub("\\(", "", QUAC.K_7[,2])
QUAC.K_7[,2] <- gsub("\\)", "", QUAC.K_7[,2])

# FUNCTION WORKSPACE BELOW ----
# readQ
readQ <- function (files = NULL, filetype = "auto", indlabfromfile = FALSE, 
                   readci = FALSE) 
{
  if (is.null(files) || (length(files) == 0)) 
    stop("readQ: No input files.")
  if (!is.character(files)) 
    stop("readQ: Argument 'files' is not a character datatype.")
  flen <- length(files)
  len <- length(files)
  dlist <- vector("list")
  for (i in seq_along(files)) {
    if (filetype == "auto") {
      chk <- tolower(checkQ(files[i])$type)
      if (chk %in% c("structure", "tess", "baps", "basic", 
                     "clumpp")) {
        if (chk == "structure") 
          dfr <- readQStructure(files[i], indlabfromfile = indlabfromfile, 
                                readci = readci)
        if (chk == "tess") 
          dfr <- readQTess(files[i])
        if (chk == "basic") 
          dfr <- readQBasic(files[i])
        if (chk == "clumpp") 
          dfr <- readQClumpp(files[i])
        if (chk == "baps") 
          dfr <- readQBaps(files[i])
        dlist <- append(dlist, dfr)
      }
      else {
        warning(paste0("readQ: Input file ", files[i], 
                       " was not identified as a STRUCTURE, TESS, BAPS, BASIC or CLUMPP filetype. Specify 'filetype' manually or check input.\n"))
      }
    }
    else {
      if (filetype == "structure") 
        dfr <- readQStructure(files[i], indlabfromfile = indlabfromfile, 
                              readci = readci)
      if (filetype == "tess") 
        dfr <- readQTess(files[i])
      if (filetype == "basic") 
        dfr <- readQBasic(files[i])
      if (filetype == "clumpp") 
        dfr <- readQClumpp(files[i])
      if (filetype == "baps") 
        dfr <- readQBaps(files[i])
      dlist <- append(dlist, dfr)
    }
  }
  return(dlist)
}

# readQStructure
readQStructure <- function (files = NULL, indlabfromfile = FALSE, readci = FALSE) 
{
  if (is.null(files) || (length(files) == 0)) 
    stop("readQStructure: No input files.")
  flen <- length(files)
  if (any(checkQ(files)$type != "STRUCTURE")) 
    stop("readQStructure: Input may be in incorrect format.")
  i <- 1
  dlist <- vector("list", length = flen)
  len1 <- length(files)
  for (i in seq_along(files)) {
    fname <- basename(files[i])
    file1 <- readLines(as.character(files[i]), warn = FALSE)
    ind <- as.numeric(as.character(gsub("\\D", "", grep("\\d individuals", 
                                                        file1, perl = TRUE, ignore.case = TRUE, value = TRUE)[1])))
    if (is.na(ind)) 
      warning(paste0("Number of individuals is NA in file: ", 
                     fname, "\n"))
    k <- as.numeric(as.character(gsub("\\D", "", grep("\\d populations assumed", 
                                                      file1, perl = TRUE, ignore.case = TRUE, value = TRUE)[1])))
    if (is.na(k)) 
      warning(paste0("Value of K is NA in file: ", fname, 
                     "\n"))
    loci <- as.numeric(gsub("\\D", "", grep("\\d loci", file1, 
                                            perl = TRUE, ignore.case = TRUE, value = TRUE)[1]))
    if (is.na(loci)) 
      warning(paste0("Number of Loci is NA in file: ", 
                     files[i], "\n"))
    burnin <- as.numeric(gsub("\\D", "", grep("\\d Burn-in period", 
                                              file1, perl = TRUE, ignore.case = TRUE, value = TRUE)[1]))
    if (is.na(burnin)) 
      warning(paste0("Burn-in value is NA in file: ", files[i], 
                     "\n"))
    reps <- as.numeric(gsub("\\D", "", grep("\\d Reps", file1, 
                                            perl = TRUE, ignore.case = TRUE, value = TRUE)[1]))
    if (is.na(reps)) 
      warning(paste0("Reps value is NA in file: ", files[i], 
                     "\n"))
    elpd <- as.numeric(gsub("=", "", gsub("Estimated Ln Prob of Data", 
                                          "", grep("Estimated Ln Prob of Data", file1, perl = TRUE, 
                                                   ignore.case = TRUE, value = TRUE)[1])))
    if (is.na(elpd)) 
      warning(paste0("Estimated Ln Prob of Data is NA in file: ", 
                     files[i], "\n"))
    mvll <- as.numeric(gsub("=", "", gsub("Mean value of ln likelihood", 
                                          "", grep("Mean value of ln likelihood", file1, perl = TRUE, 
                                                   ignore.case = TRUE, value = TRUE)[1])))
    if (is.na(mvll)) 
      warning(paste0("Mean value of ln likelihood is NA in file: ", 
                     files[i], "\n"))
    vll <- as.numeric(gsub("=", "", gsub("Variance of ln likelihood", 
                                         "", grep("Variance of ln likelihood", file1, perl = TRUE, 
                                                  ignore.case = TRUE, value = TRUE)[1])))
    if (is.na(vll)) 
      warning(paste0("Variance of ln likelihood is NA in file: ", 
                     files[i], "\n"))
    file1 <- file1[grep(".+\\(\\d+\\).+\\:.+", file1)]
    if (length(file1) == 0) {
      cstart <- charmatch("Inferred ancestry of individuals", 
                          file1)
      cend <- charmatch("Estimated Allele Frequencies in each", 
                        file1)
      file1 <- file1[(cstart + 2):(cend - 1)]
    }
    file_a <- file1[file1 != ""]
    rm(file1)
    tc_file_a <- textConnection(file_a)
    file_b <- read.delim(tc_file_a, header = FALSE, sep = "", 
                         stringsAsFactors = FALSE)
    close(tc_file_a)
    suppressWarnings(errorcheck <- try(file_b[1, as.integer(grep(":", 
                                                                 file_b[1, ]) + 1):as.integer(max(grep("^[0-9]|[.]+$", 
                                                                                                       file_b[1, ]))), drop = FALSE], silent = TRUE))
    rm(file_b)
    if (class(errorcheck) == "try-error") {
      file_a <- gsub("\\([0-9.,]+\\)", "", file_a)
      file_b <- gsub(":  ", "", substr(file_a, regexpr(":\\W+\\d\\.\\d+", 
                                                       file_a), nchar(file_a) - 1))
      file_b <- sub("\\s+$", "", sub("^\\s+", "", file_b))
      rm(file_a)
      file_c <- as.vector(as.numeric(as.character(unlist(strsplit(file_b, 
                                                                  " ")))))
      rm(file_b)
      dframe <- as.data.frame(matrix(file_c, nrow = ind, 
                                     byrow = TRUE), stringsAsFactors = FALSE)
    }
    else {
      tc_file_a <- textConnection(file_a)
      file_b <- read.delim(tc_file_a, header = FALSE, sep = "", 
                           stringsAsFactors = FALSE)
      close(tc_file_a)
      dframe <- file_b[, as.integer(grep(":", file_b[1, 
      ]) + 1):as.integer(max(grep("^[0-9]|[.]+$", file_b[1, 
      ]))), drop = FALSE]
    }
    dframe <- as.data.frame(sapply(dframe, as.numeric), stringsAsFactors = FALSE)
    colnames(dframe) <- paste0("Cluster", 1:ncol(dframe))
    row.names(dframe) <- 1:nrow(dframe)
    if (indlabfromfile) {
      labeldf <- file_b[, (grep("[0-9]", file_b[1, ])[1] + 
                             1):(grep("[(]", file_b[1, ])[1] - 1), drop = FALSE]
      if (ncol(labeldf) > 1) 
        labeldf <- data.frame(V2 = do.call(paste, c(labeldf, 
                                                    sep = "_")), stringsAsFactors = FALSE)
      if (nrow(labeldf) == nrow(dframe)) {
        if (any(duplicated(labeldf[, 1]))) {
          warning(paste0("readQStructure: Individual names in file ", 
                         fname, " not used due to presence of duplicate names.\n"))
        }
        else {
          row.names(dframe) <- as.character(labeldf[, 
                                                    1])
        }
      }
      else {
        warning(paste0("readQStructure: Individual names in file ", 
                       fname, " not used due to incorrect length.\n"))
      }
    }
    attr(dframe, "ind") <- nrow(dframe)
    attr(dframe, "k") <- ncol(dframe)
    attr(dframe, "loci") <- loci
    attr(dframe, "burnin") <- burnin
    attr(dframe, "reps") <- reps
    attr(dframe, "elpd") <- elpd
    attr(dframe, "mvll") <- mvll
    attr(dframe, "vll") <- vll
    if (readci) {
      cichk <- grep("([0-9.]+,[0-9.]+)", file_b[1, ])
      if (length(cichk) != 0) {
        file_b <- apply(file_b[, cichk, drop = FALSE], 
                        1, paste0, collapse = "")
        file_b <- gsub("[()]", "", gsub(")(", ",", file_b, 
                                        fixed = TRUE))
        cframe <- as.data.frame(matrix(as.numeric(unlist(strsplit(file_b, 
                                                                  ","))), ncol = ncol(dframe) * 2, byrow = TRUE))
        colnames(cframe) <- as.vector(t(outer(paste0("Cluster", 
                                                     1:ncol(dframe)), c("L", "H"), paste, sep = "")))
        row.names(cframe) <- row.names(dframe)
        attr(dframe, "ci") <- cframe
      }
      else {
        warning("plotQStructure: Confidence intervals could not be read.\n")
      }
    }
    dlist[[i]] <- dframe
  }
  fnames <- sub(".txt", "", basename(files))
  names(dlist) <- fnames
  return(dlist)
}
