# %%%%%%%%%%%%%%%%%%
# %%% RESAMPLING %%%
# %%%%%%%%%%%%%%%%%%

# This script generates Resampling arrays, and plots their results, 
# for Quercus acerifolia (QUAC) and Quercus boyntonii (QUBO).
# For both species, both the optimized de novo assemblies (DNFA) and the reference analyses (REF)
# are analyzed, at both the R0 and the R80 levels of missing data. In addition to processing 
# SNP datasets (described above), this script also  reads in the QUAC and QUBO 
# microsatellite (MSAT) genind files as well. It analyzes both the complete SNP and MSAT datasets 
# AND each dataset subset to the samples shared between the SNP and MSAT studies 
# (for a more direct comparison).

# For each species, 2 MSAT datasets are examined: Complete and Subset
# For each species, 8 SNP datasets are examined: De novo and Reference x R0 and R80 x Complete and Subset
# Therefore, for each species, 10 datasets are examined, meaning 20 datasets are examined in total

# Resampling arrays are collections of values indicating the percentage of total wild alleles
# that are represented by a (random) subset of wild samples. Each column of these arrays corresponds
# to the category of allelic representation of the random subset of wild samples (total allelic diversity,
# very common alleles, common alleles, low frequency alleles, and rare alleles).
# After resampling arrays are created, plots are generated representing the resampling analyses.

# For QUAC, SNP and MSAT subsets are generated by using the Tissue database names for each SNP/MSAT sample
# For QUBO, SNP and MSAT subsets are generated by shortening names to only IMLS sample names

# Resampling arrays are saved as .Rdata objects, and plots are saved as .png files, 
# to enable running this script outside of RStudio.

library(adegenet)
library(RColorBrewer)
library(scales)
library(parallel)

# %%%% FUNCTIONS, VARIABLES, AND PARALLELIZATION %%%% ----
# Read in relevant functions required for resampling analyses
SSRvSNP.wd <- "~/Documents/SSRvSNP/Code/"
setwd(SSRvSNP.wd)
source("exSituRepresentation/functions_exSituRepresentation.R")
# Declare a directory within which to store .Rdata objects of resampling arrays, and plots
resamplingDataDir <- paste0(SSRvSNP.wd, "exSituRepresentation/Resampling/resamplingData/")
# Dimensions for resmpling plots
plotWidth <- 1262 ; plotHeight <- 734
# Plotting colors (for all plots!), with transparency for values other than Total (alpha = 1.0 is transparent)
plotColors <- c("red","red4","darkorange3","coral","purple")
plotColors[2:5] <- alpha(plotColors[2:5], 0.6)
# Set up relevant cores, and make sure adegenet library is present on cluster
num_cores <- detectCores() - 20 ; cl <- makeCluster(num_cores)
clusterEvalQ(cl, library("adegenet"))
# Specify number of replicates. This value is used uniformly, for ALL datasets (QUAC and QUBO)
num_reps <- 5000
# Export relevant functions and variables
clusterExport(cl, varlist = c("getAlleleCategories", "exSitu_Sample", "exSitu_Resample", "num_reps"))

# %%%% QUAC %%%% ----
# ---- MSATS (COMPLETE) ----
# READ IN GENIND FILE (GCC_QUAC_ZAIN repo; QUAC_woK_garden_wild_clean.gen)
QUAC.MSAT.filePath <- 
  "~/Documents/peripheralProjects/GCC_QUAC_ZAIN/Data_Files/"
QUAC.MSAT.genind <- 
  read.genepop(paste0(QUAC.MSAT.filePath, "Adegenet_Files/QUAC_woK_allpop_clean.gen"), ncode = 3)
# Assign sample names: read in Tissue database names from GCC_QUAC_ZAIN repository
QUAC.MSAT.tissueNames <- 
  unlist(read.csv2(paste0(QUAC.MSAT.filePath, "Data_Frames/QUAC_woK_allpop_clean_df.csv"), header = TRUE, sep=",")[1])
rownames(QUAC.MSAT.genind@tab) <- QUAC.MSAT.tissueNames
# Correct popNames: samples with popName pattern QAc-G- are garden 
levels(QUAC.MSAT.genind@pop)[grep(pattern = "QAc-G-", levels(QUAC.MSAT.genind@pop))] <- 
  rep("garden", length(grep(pattern = "QAc-G-", levels(QUAC.MSAT.genind@pop))))
# Correct popNames: samples with popName pattern QAc-W- are wild
levels(QUAC.MSAT.genind@pop)[grep(pattern = "QAc-W-", levels(QUAC.MSAT.genind@pop))] <- 
  rep("wild", length(grep(pattern = "QAc-W-", levels(QUAC.MSAT.genind@pop))))

# CREATE RESAMPLING ARRAY AND CALCULATE SUMMARY STATISTICS
# Export genind object
clusterExport(cl, varlist = "QUAC.MSAT.genind")
# Run resampling in parallel, to generate an array that's saved to disc
samplingResults_QUAC.MSAT <- 
  exSitu_Resample_Parallel(QUAC.MSAT.genind, cluster = cl, reps = num_reps,
                           arrayFilepath=paste0(resamplingDataDir,"QUAC.MSAT.Complete_resampArr.Rdata"))
# Close cores
# stopCluster(cl)

# Average results across replicates (slices) of the sampling array, to determine
# the minimum number of samples required to capture 95% wild genetic diversity
# (We average samplingResults[,1,], since this column contains the total genetic diversity)
# Standard deviation is calculated as well
min_95_QUAC.MSAT <- min(which(apply(samplingResults_QUAC.MSAT[,1,],1,mean) > 95))
min_95.SD_QUAC.MSAT <- apply(samplingResults_QUAC.MSAT[,1,],1,sd)[min_95_QUAC.MSAT]
print(min_95_QUAC.MSAT); print(min_95.SD_QUAC.MSAT)
# Calculate means, for each capture rate category
total_means <- apply(samplingResults_QUAC.MSAT[,1,], 1, mean)
v.com_means <- apply(samplingResults_QUAC.MSAT[,2,], 1, mean)
com_means <- apply(samplingResults_QUAC.MSAT[,3,], 1, mean)
lowfr_means <- apply(samplingResults_QUAC.MSAT[,4,], 1, mean)
rare_means <- apply(samplingResults_QUAC.MSAT[,5,], 1, mean)

# *** PLOTTING ----
# Call png command, to save resampling plots to disk
png(file = paste0(resamplingDataDir, "QUAC.MSAT.Complete.png"), width = plotWidth, height = plotHeight)
# Plots all sets of points onto single graph, as well as 95% threshold line
plot(total_means, ylim=c(0,110), col=plotColors[1], pch=16,
     xlab="Number of Individuals", ylab="Percent Diversity Capture",
     main=paste0("QUAC, Microsatellites (Complete: 164 samples; ", num_reps, " Replicates)"))
points(v.com_means, col=plotColors[2], pch=16)
points(com_means, col=plotColors[3], pch=16)
points(lowfr_means, col=plotColors[4], pch=16)
points(rare_means, col=plotColors[5], pch=16)
legend(x=140, y=70.13276, inset = 0.05, legend = c("Total","Very common","Common","Low frequency", "Rare"),
       col=plotColors, pch = c(20,20,20), cex=1, pt.cex = 2, bty="n", y.intersp = 1)
# Lines for 95% threshold
abline(h=95, col="black", lty=3); abline(v=min_95_QUAC.MSAT, col="black")
# Text for number of individuals to capture 95% threshold, with position based on variable
mtext(text=paste0("Minimum sampling size (95%) = ", min_95_QUAC.MSAT), 
      side=1, line=-1.5, at=min_95_QUAC.MSAT-20)
# Turn off plotting device
dev.off()

# ---- SNPS (COMPLETE) ----
# %%% DE NOVO ----
# R0 ----
# READ IN GENIND FILE (QUAC DNFA; R0, min-maf=0; first SNP per locus; 2 populations, garden and wild)
genpop.filePath <- 
  "/RAID1/IMLS_GCCO/Analysis/Stacks/denovo_finalAssemblies/QUAC/output/populations_R0_NOMAF_1SNP_2Pops_NoK/"
QUAC.SNP.DN.R0.genind <- read.genepop(paste0(genpop.filePath,"populations.snps.gen"))
# Correct popNames
pop(QUAC.SNP.DN.R0.genind) <- 
  factor(read.table(paste0(genpop.filePath, "QUAC_popmap_GardenWild_NoK"), header=FALSE)[,2])
# Get QUAC SNP Tissue sample names, and rename SNP genind matrix
# File was created (by Austin K.), and can be found on Hoban Lab Drive ("MSATcomparisons_TissueNames")
QUAC.SNP.DN.R0.tissueNames_filepath <- paste0(SSRvSNP.wd,"exSituRepresentation/QUAC_SNP_TissueNames.csv")
QUAC.SNP.DN.R0.tissueNames <- unlist(read.csv2(QUAC.SNP.DN.R0.tissueNames_filepath, header = TRUE, sep = ",")[3])
rownames(QUAC.SNP.DN.R0.genind@tab) <- QUAC.SNP.DN.R0.tissueNames

# CREATE RESAMPLING ARRAY AND CALCULATE SUMMARY STATISTICS
# Export genind object
clusterExport(cl, varlist = "QUAC.SNP.DN.R0.genind")
# Run resampling in parallel, to generate an array that's saved to disc
samplingResults_QUAC.SNP.DN.R0 <- 
  exSitu_Resample_Parallel(QUAC.SNP.DN.R0.genind, cluster = cl, reps = num_reps,
                           arrayFilepath=paste0(resamplingDataDir,"QUAC.SNP.DN.R0.Complete_resampArr.Rdata"))
# Close cores
# stopCluster(cl)

# Average 95% minimum sampling size across replicates, as well as standard deviation
min_95_QUAC.SNP.DN.R0 <- min(which(apply(samplingResults_QUAC.SNP.DN.R0[,1,],1,mean) > 95))
min_95.SD_QUAC.SNP.DN.R0 <- apply(samplingResults_QUAC.SNP.DN.R0[,1,],1,sd)[min_95_QUAC.SNP.DN.R0]
print(min_95_QUAC.SNP.DN.R0); print(min_95.SD_QUAC.SNP.DN.R0)
# Calculate means, for each capture rate category
total_means <- apply(samplingResults_QUAC.SNP.DN.R0[,1,], 1, mean)
v.com_means <- apply(samplingResults_QUAC.SNP.DN.R0[,2,], 1, mean)
com_means <- apply(samplingResults_QUAC.SNP.DN.R0[,3,], 1, mean)
lowfr_means <- apply(samplingResults_QUAC.SNP.DN.R0[,4,], 1, mean)
rare_means <- apply(samplingResults_QUAC.SNP.DN.R0[,5,], 1, mean)

# R80 ----
# READ IN GENIND FILE (QUAC DNFA; R80, min-maf=0; first SNP per locus; 2 populations, garden and wild)
genpop.filePath <- 
  "/RAID1/IMLS_GCCO/Analysis/Stacks/denovo_finalAssemblies/QUAC/output/populations_R80_NOMAF_1SNP_2Pops_NoK/"
QUAC.SNP.DN.R80.genind <- read.genepop(paste0(genpop.filePath,"populations.snps.gen"))
# Correct popNames
pop(QUAC.SNP.DN.R80.genind) <- 
  factor(read.table(paste0(genpop.filePath, "QUAC_popmap_GardenWild_NoK"), header=FALSE)[,2])
# Get QUAC SNP Tissue sample names, and rename SNP genind matrix
# File was created (by Austin K.), and can be found on Hoban Lab Drive ("MSATcomparisons_TissueNames")
QUAC.SNP.DN.R80.tissueNames_filepath <- paste0(SSRvSNP.wd,"exSituRepresentation/QUAC_SNP_TissueNames.csv")
QUAC.SNP.DN.R80.tissueNames <- unlist(read.csv2(QUAC.SNP.DN.R80.tissueNames_filepath, header = TRUE, sep = ",")[3])
rownames(QUAC.SNP.DN.R80.genind@tab) <- QUAC.SNP.DN.R80.tissueNames

# CREATE RESAMPLING ARRAY AND CALCULATE SUMMARY STATISTICS
# Export genind object
clusterExport(cl, varlist = "QUAC.SNP.DN.R80.genind")
# Run resampling in parallel, to generate an array that's saved to disc
samplingResults_QUAC.SNP.DN.R80 <- 
  exSitu_Resample_Parallel(QUAC.SNP.DN.R80.genind, cluster = cl, reps = num_reps,
                           arrayFilepath=paste0(resamplingDataDir,"QUAC.SNP.DN.R80.Complete_resampArr.Rdata"))
# Close cores
# stopCluster(cl)

# Average 95% minimum sampling size across replicates, as well as standard deviation
min_95_QUAC.SNP.DN.R80 <- min(which(apply(samplingResults_QUAC.SNP.DN.R80[,1,],1,mean) > 95))
min_95.SD_QUAC.SNP.DN.R80 <- apply(samplingResults_QUAC.SNP.DN.R80[,1,],1,sd)[min_95_QUAC.SNP.DN.R80]
print(min_95_QUAC.SNP.DN.R80); print(min_95.SD_QUAC.SNP.DN.R80)
# Calculate means, for each capture rate category
total_means <- apply(samplingResults_QUAC.SNP.DN.R80[,1,], 1, mean)
v.com_means <- apply(samplingResults_QUAC.SNP.DN.R80[,2,], 1, mean)
com_means <- apply(samplingResults_QUAC.SNP.DN.R80[,3,], 1, mean)
lowfr_means <- apply(samplingResults_QUAC.SNP.DN.R80[,4,], 1, mean)
rare_means <- apply(samplingResults_QUAC.SNP.DN.R80[,5,], 1, mean)

# *** PLOTTING ----
# Call png command, to save resampling plots to disk
png(file = paste0(resamplingDataDir, "QUAC.SNP.DN.R0-R80.Complete.png"), width = plotWidth, height = plotHeight)
# Set plotting window to stack 2 graphs vertically
par(mfcol=c(2,1), oma=rep(0.2,4))
# R0
# Plots all sets of points onto single graph, as well as 95% threshold line
plot(total_means, ylim=c(0,110), col=plotColors[1], pch=16,
     xlab="Number of Individuals", ylab="Percent Diversity Capture",
     main=paste0("QUAC, SNPs: De novo, R0 (Complete: 91 samples; ", num_reps, " Replicates)"))
points(v.com_means, col=plotColors[2], pch=16)
points(com_means, col=plotColors[3], pch=16)
points(lowfr_means, col=plotColors[4], pch=16)
points(rare_means, col=plotColors[5], pch=16)
legend(x=83, y=70.13276, inset = 0.05, legend = c("Total","Very common","Common","Low frequency", "Rare"),
       col=plotColors, pch = c(20,20,20), cex=1, pt.cex = 2, bty="n", y.intersp = 1)
# Lines for 95% threshold
abline(h=95, col="black", lty=3); abline(v=min_95_QUAC.SNP.DN.R0, col="black")
# Text for number of individuals to capture 95% threshold, with position based on variable
mtext(text=paste0("Minimum sampling size (95%) = ", min_95_QUAC.SNP.DN.R0), 
      side=1, line=-1.5, at=min_95_QUAC.SNP.DN.R0-10)
# R80
# Plots all sets of points onto single graph, as well as 95% threshold line
plot(total_means, ylim=c(0,110), col=plotColors[1], pch=16,
     xlab="Number of Individuals", ylab="Percent Diversity Capture",
     main=paste0("QUAC, SNPs: De novo, R80 (Complete: 91 samples; ", num_reps, " Replicates)"))
points(v.com_means, col=plotColors[2], pch=16)
points(com_means, col=plotColors[3], pch=16)
points(lowfr_means, col=plotColors[4], pch=16)
points(rare_means, col=plotColors[5], pch=16)
legend(x=83, y=70.13276, inset = 0.05, legend = c("Total","Very common","Common","Low frequency", "Rare"),
       col=plotColors, pch = c(20,20,20), cex=1, pt.cex = 2, bty="n", y.intersp = 1)
# Lines for 95% threshold
abline(h=95, col="black", lty=3); abline(v=min_95_QUAC.SNP.DN.R80, col="black")
# Text for number of individuals to capture 95% threshold, with position based on variable
mtext(text=paste0("Minimum sampling size (95%) = ", min_95_QUAC.SNP.DN.R80), 
      side=1, line=-1.5, at=min_95_QUAC.SNP.DN.R80-10)
# Turn off plotting device
dev.off()

# %%% REFERENCE ----
# R0 ----
# READ IN GENIND FILE: QUAC GSNAP4 alignment; R0, min-maf=0, first SNP/locus, 2 populations (garden and wild)
genpop.filePath <- 
  "/RAID1/IMLS_GCCO/Analysis/Stacks/reference_filteredReads/QUAC/Q_rubra/output/populations_R0_NOMAF_1SNP_2Pops_NoK/"
QUAC.SNP.REF.R0.genind <- read.genepop(paste0(genpop.filePath,"populations.snps.gen"))
# Correct popNames
pop(QUAC.SNP.REF.R0.genind) <- 
  factor(read.table(paste0(genpop.filePath, "QUAC_popmap_GardenWild_NoK"), header=FALSE)[,2])
# Get QUAC SNP Tissue sample names, and rename SNP genind matrix
# File was created (by Austin K.), and can be found on Hoban Lab Drive ("MSATcomparisons_TissueNames")
QUAC.SNP.REF.R0.tissueNames_filepath <- paste0(SSRvSNP.wd,"exSituRepresentation/QUAC_SNP_TissueNames.csv")
QUAC.SNP.REF.R0.tissueNames <- unlist(read.csv2(QUAC.SNP.REF.R0.tissueNames_filepath, header = TRUE, sep = ",")[3])
rownames(QUAC.SNP.REF.R0.genind@tab) <- QUAC.SNP.REF.R0.tissueNames

# CREATE RESAMPLING ARRAY AND CALCULATE SUMMARY STATISTICS
# Export genind object
clusterExport(cl, varlist = "QUAC.SNP.REF.R0.genind")
# Run resampling in parallel, to generate an array that's saved to disc
samplingResults_QUAC.SNP.REF.R0 <- 
  exSitu_Resample_Parallel(QUAC.SNP.REF.R0.genind, cluster = cl, reps = num_reps,
                           arrayFilepath=paste0(resamplingDataDir,"QUAC.SNP.REF.R0.Complete_resampArr.Rdata"))
# Close cores
# stopCluster(cl)

# Average 95% minimum sampling size across replicates, as well as standard deviation
min_95_QUAC.SNP.REF.R0 <- min(which(apply(samplingResults_QUAC.SNP.REF.R0[,1,],1,mean) > 95))
min_95.SD_QUAC.SNP.REF.R0 <- apply(samplingResults_QUAC.SNP.REF.R0[,1,],1,sd)[min_95_QUAC.SNP.REF.R0]
print(min_95_QUAC.SNP.REF.R0); print(min_95.SD_QUAC.SNP.REF.R0)
# Calculate means, for each capture rate category
total_means <- apply(samplingResults_QUAC.SNP.REF.R0[,1,], 1, mean)
v.com_means <- apply(samplingResults_QUAC.SNP.REF.R0[,2,], 1, mean)
com_means <- apply(samplingResults_QUAC.SNP.REF.R0[,3,], 1, mean)
lowfr_means <- apply(samplingResults_QUAC.SNP.REF.R0[,4,], 1, mean)
rare_means <- apply(samplingResults_QUAC.SNP.REF.R0[,5,], 1, mean)

# R80 ----
# READ IN GENIND FILE: QUAC GSNAP4 alignment; R80, min-maf=0, first SNP/locus, 2 populations (garden and wild)
genpop.filePath <- 
  "/RAID1/IMLS_GCCO/Analysis/Stacks/reference_filteredReads/QUAC/Q_rubra/output/populations_R80_NOMAF_1SNP_2Pops_NoK/"
QUAC.SNP.REF.R80.genind <- read.genepop(paste0(genpop.filePath,"populations.snps.gen"))
# Correct popNames
pop(QUAC.SNP.REF.R80.genind) <- 
  factor(read.table(paste0(genpop.filePath, "QUAC_popmap_GardenWild_NoK"), header=FALSE)[,2])
# Get QUAC SNP Tissue sample names, and rename SNP genind matrix
# File was created (by Austin K.), and can be found on Hoban Lab Drive ("MSATcomparisons_TissueNames")
QUAC.SNP.REF.R80.tissueNames_filepath <- paste0(SSRvSNP.wd,"exSituRepresentation/QUAC_SNP_TissueNames.csv")
QUAC.SNP.REF.R80.tissueNames <- unlist(read.csv2(QUAC.SNP.REF.R80.tissueNames_filepath, header = TRUE, sep = ",")[3])
rownames(QUAC.SNP.REF.R80.genind@tab) <- QUAC.SNP.REF.R80.tissueNames

# CREATE RESAMPLING ARRAY AND CALCULATE SUMMARY STATISTICS
# Export genind object
clusterExport(cl, varlist = "QUAC.SNP.REF.R80.genind")
# Run resampling in parallel, to generate an array that's saved to disc
samplingResults_QUAC.SNP.REF.R80 <- 
  exSitu_Resample_Parallel(QUAC.SNP.REF.R80.genind, cluster = cl, reps = num_reps,
                           arrayFilepath=paste0(resamplingDataDir,"QUAC.SNP.REF.R80.Complete_resampArr.Rdata"))
# Close cores
# stopCluster(cl)

# Average 95% minimum sampling size across replicates, as well as standard deviation
min_95_QUAC.SNP.REF.R80 <- min(which(apply(samplingResults_QUAC.SNP.REF.R80[,1,],1,mean) > 95))
min_95.SD_QUAC.SNP.REF.R80 <- apply(samplingResults_QUAC.SNP.REF.R80[,1,],1,sd)[min_95_QUAC.SNP.REF.R80]
print(min_95_QUAC.SNP.REF.R80); print(min_95.SD_QUAC.SNP.REF.R80)
# Calculate means, for each capture rate category
total_means <- apply(samplingResults_QUAC.SNP.REF.R80[,1,], 1, mean)
v.com_means <- apply(samplingResults_QUAC.SNP.REF.R80[,2,], 1, mean)
com_means <- apply(samplingResults_QUAC.SNP.REF.R80[,3,], 1, mean)
lowfr_means <- apply(samplingResults_QUAC.SNP.REF.R80[,4,], 1, mean)
rare_means <- apply(samplingResults_QUAC.SNP.REF.R80[,5,], 1, mean)

# *** PLOTTING ----
# Call png command, to save resampling plots to disk
png(file = paste0(resamplingDataDir, "QUAC.SNP.REF.R0-R80.Complete.png"), width = plotWidth, height = plotHeight)
# Set plotting window to stack 2 graphs vertically
par(mfcol=c(2,1), oma=rep(0.2,4))
# R0
# Plots all sets of points onto single graph, as well as 95% threshold line
plot(total_means, ylim=c(0,110), col=plotColors[1], pch=16,
     xlab="Number of Individuals", ylab="Percent Diversity Capture",
     main=paste0("QUAC, SNPs: Reference, R0 (Complete: 91 samples; ", num_reps, " Replicates)"))
points(v.com_means, col=plotColors[2], pch=16)
points(com_means, col=plotColors[3], pch=16)
points(lowfr_means, col=plotColors[4], pch=16)
points(rare_means, col=plotColors[5], pch=16)
legend(x=83, y=70.13276, inset = 0.05, legend = c("Total","Very common","Common","Low frequency", "Rare"),
       col=plotColors, pch = c(20,20,20), cex=1, pt.cex = 2, bty="n", y.intersp = 1)
# Lines for 95% threshold
abline(h=95, col="black", lty=3); abline(v=min_95_QUAC.SNP.REF.R0, col="black")
# Text for number of individuals to capture 95% threshold, with position based on variable
mtext(text=paste0("Minimum sampling size (95%) = ", min_95_QUAC.SNP.REF.R0), 
      side=1, line=-1.5, at=min_95_QUAC.SNP.REF.R0-10)
# R80
# Plots all sets of points onto single graph, as well as 95% threshold line
plot(total_means, ylim=c(0,110), col=plotColors[1], pch=16,
     xlab="Number of Individuals", ylab="Percent Diversity Capture",
     main=paste0("QUAC, SNPs: Reference, R80 (Complete: 91 samples; ", num_reps, " Replicates)"))
points(v.com_means, col=plotColors[2], pch=16)
points(com_means, col=plotColors[3], pch=16)
points(lowfr_means, col=plotColors[4], pch=16)
points(rare_means, col=plotColors[5], pch=16)
legend(x=83, y=70.13276, inset = 0.05, legend = c("Total","Very common","Common","Low frequency", "Rare"),
       col=plotColors, pch = c(20,20,20), cex=1, pt.cex = 2, bty="n", y.intersp = 1)
# Lines for 95% threshold
abline(h=95, col="black", lty=3); abline(v=min_95_QUAC.SNP.REF.R80, col="black")
# Text for number of individuals to capture 95% threshold, with position based on variable
mtext(text=paste0("Minimum sampling size (95%) = ", min_95_QUAC.SNP.REF.R80), 
      side=1, line=-1.5, at=min_95_QUAC.SNP.REF.R80-10)
# Turn off plotting device
dev.off()

# ---- MSATS AND SNPS: SUBSET ----
# Subset SNP sample names by those that are also seen within the MSAT samples 
QUAC_sharedSamples <- sort(QUAC.SNP.DN.R0.tissueNames[which(QUAC.SNP.DN.R0.tissueNames %in% QUAC.MSAT.tissueNames)])
# Subset MSAT and SNP genind matrices to strictly shared samples, dropping now absent alleles
QUAC.MSAT_subset.genind <- QUAC.MSAT.genind[QUAC_sharedSamples,, drop=TRUE]
# De novo
QUAC.SNP.DN.R0_subset.genind <- QUAC.SNP.DN.R0.genind[QUAC_sharedSamples,, drop=TRUE]
QUAC.SNP.DN.R80_subset.genind <- QUAC.SNP.DN.R80.genind[QUAC_sharedSamples,, drop=TRUE]
# Reference
QUAC.SNP.REF.R0_subset.genind <- QUAC.SNP.REF.R0.genind[QUAC_sharedSamples,, drop=TRUE]
QUAC.SNP.REF.R80_subset.genind <- QUAC.SNP.REF.R80.genind[QUAC_sharedSamples,, drop=TRUE]

# CREATE RESAMPLING ARRAY AND CALCULATE SUMMARY STATISTICS
# Export genind objects
clusterExport(cl, varlist = c("QUAC.MSAT_subset.genind", "QUAC.SNP.DN.R0_subset.genind",
                              "QUAC.SNP.DN.R80_subset.genind", "QUAC.SNP.REF.R0_subset.genind", 
                              "QUAC.SNP.REF.R80_subset.genind"))
# Run resampling in parallel, to generate arrays
# MSAT
samplingResults_QUAC.MSAT_subset <- 
  exSitu_Resample_Parallel(QUAC.MSAT_subset.genind, cluster = cl, reps = num_reps,
                           arrayFilepath=paste0(resamplingDataDir,"QUAC.MSAT.Subset_resampArr.Rdata"))
# SNP: De novo
# R0
samplingResults_QUAC.SNP.DN.R0_subset <- 
  exSitu_Resample_Parallel(QUAC.SNP.DN.R0_subset.genind, cluster = cl, reps = num_reps,
                           arrayFilepath=paste0(resamplingDataDir,"QUAC.SNP.DN.R0.Subset_resampArr.Rdata"))
# R80
samplingResults_QUAC.SNP.DN.R80_subset <- 
  exSitu_Resample_Parallel(QUAC.SNP.DN.R80_subset.genind, cluster = cl, reps = num_reps,
                           arrayFilepath=paste0(resamplingDataDir,"QUAC.SNP.DN.R80.Subset_resampArr.Rdata"))
# SNP: Reference
# R0
samplingResults_QUAC.SNP.REF.R0_subset <- 
  exSitu_Resample_Parallel(QUAC.SNP.REF.R0_subset.genind, cluster = cl, reps = num_reps,
                           arrayFilepath=paste0(resamplingDataDir,"QUAC.SNP.REF.R0.Subset_resampArr.Rdata"))
# R80
samplingResults_QUAC.SNP.REF.R80_subset <- 
  exSitu_Resample_Parallel(QUAC.SNP.REF.R80_subset.genind, cluster = cl, reps = num_reps,
                           arrayFilepath=paste0(resamplingDataDir,"QUAC.SNP.REF.R80.Subset_resampArr.Rdata"))
# Close cores
# stopCluster(cl)

# MSAT_subset: average 95% minimum sampling size across replicates, as well as standard deviation
min_95_QUAC.MSAT_subset <- min(which(apply(samplingResults_QUAC.MSAT_subset[,1,],1,mean) > 95))
min_95.SD_QUAC.MSAT_subset <- apply(samplingResults_QUAC.MSAT_subset[,1,],1,sd)[min_95_QUAC.MSAT_subset]
print(min_95_QUAC.MSAT_subset); print(min_95.SD_QUAC.MSAT_subset)
# Calculate means, for each capture rate category
total_means <- apply(samplingResults_QUAC.MSAT_subset[,1,], 1, mean)
v.com_means <- apply(samplingResults_QUAC.MSAT_subset[,2,], 1, mean)
com_means <- apply(samplingResults_QUAC.MSAT_subset[,3,], 1, mean)
lowfr_means <- apply(samplingResults_QUAC.MSAT_subset[,4,], 1, mean)
rare_means <- apply(samplingResults_QUAC.MSAT_subset[,5,], 1, mean)

# SNP_subset: average 95% minimum sampling size across replicates, as well as standard deviation
# De novo
# R0
min_95_QUAC.SNP.DN.R0_subset <- min(which(apply(samplingResults_QUAC.SNP.DN.R0_subset[,1,],1,mean) > 95))
min_95.SD_QUAC.SNP.DN.R0_subset <- apply(samplingResults_QUAC.SNP.DN.R0_subset[,1,],1,sd)[min_95_QUAC.SNP.DN.R0_subset]
print(min_95_QUAC.SNP.DN.R0_subset); print(min_95.SD_QUAC.SNP.DN.R0_subset)
# Calculate means, for each capture rate category
total_means <- apply(samplingResults_QUAC.SNP.DN.R0_subset[,1,], 1, mean)
v.com_means <- apply(samplingResults_QUAC.SNP.DN.R0_subset[,2,], 1, mean)
com_means <- apply(samplingResults_QUAC.SNP.DN.R0_subset[,3,], 1, mean)
lowfr_means <- apply(samplingResults_QUAC.SNP.DN.R0_subset[,4,], 1, mean)
rare_means <- apply(samplingResults_QUAC.SNP.DN.R0_subset[,5,], 1, mean)
# R80
min_95_QUAC.SNP.DN.R80_subset <- min(which(apply(samplingResults_QUAC.SNP.DN.R80_subset[,1,],1,mean) > 95))
min_95.SD_QUAC.SNP.DN.R80_subset <- apply(samplingResults_QUAC.SNP.DN.R80_subset[,1,],1,sd)[min_95_QUAC.SNP.DN.R80_subset]
print(min_95_QUAC.SNP.DN.R80_subset); print(min_95.SD_QUAC.SNP.DN.R80_subset)
# Calculate means, for each capture rate category
total_means <- apply(samplingResults_QUAC.SNP.DN.R80_subset[,1,], 1, mean)
v.com_means <- apply(samplingResults_QUAC.SNP.DN.R80_subset[,2,], 1, mean)
com_means <- apply(samplingResults_QUAC.SNP.DN.R80_subset[,3,], 1, mean)
lowfr_means <- apply(samplingResults_QUAC.SNP.DN.R80_subset[,4,], 1, mean)
rare_means <- apply(samplingResults_QUAC.SNP.DN.R80_subset[,5,], 1, mean)

# Reference
# R0
min_95_QUAC.SNP.REF.R0_subset <- min(which(apply(samplingResults_QUAC.SNP.REF.R0_subset[,1,],1,mean) > 95))
min_95.SD_QUAC.SNP.REF.R0_subset <- apply(samplingResults_QUAC.SNP.REF.R0_subset[,1,],1,sd)[min_95_QUAC.SNP.REF.R0_subset]
print(min_95_QUAC.SNP.REF.R0_subset); print(min_95.SD_QUAC.SNP.REF.R0_subset)
# Calculate means, for each capture rate category
total_means <- apply(samplingResults_QUAC.SNP.REF.R0_subset[,1,], 1, mean)
v.com_means <- apply(samplingResults_QUAC.SNP.REF.R0_subset[,2,], 1, mean)
com_means <- apply(samplingResults_QUAC.SNP.REF.R0_subset[,3,], 1, mean)
lowfr_means <- apply(samplingResults_QUAC.SNP.REF.R0_subset[,4,], 1, mean)
rare_means <- apply(samplingResults_QUAC.SNP.REF.R0_subset[,5,], 1, mean)
# R80
min_95_QUAC.SNP.REF.R80_subset <- min(which(apply(samplingResults_QUAC.SNP.REF.R80_subset[,1,],1,mean) > 95))
min_95.SD_QUAC.SNP.REF.R80_subset <- apply(samplingResults_QUAC.SNP.REF.R80_subset[,1,],1,sd)[min_95_QUAC.SNP.REF.R80_subset]
print(min_95_QUAC.SNP.REF.R80_subset); print(min_95.SD_QUAC.SNP.REF.R80_subset)
# Calculate means, for each capture rate category
total_means <- apply(samplingResults_QUAC.SNP.REF.R80_subset[,1,], 1, mean)
v.com_means <- apply(samplingResults_QUAC.SNP.REF.R80_subset[,2,], 1, mean)
com_means <- apply(samplingResults_QUAC.SNP.REF.R80_subset[,3,], 1, mean)
lowfr_means <- apply(samplingResults_QUAC.SNP.REF.R80_subset[,4,], 1, mean)
rare_means <- apply(samplingResults_QUAC.SNP.REF.R80_subset[,5,], 1, mean)

# *** PLOTTING ----
# Call png command, to save resampling plots to disk
png(file = paste0(resamplingDataDir, "QUAC.MSAT.Subset.png"), width = plotWidth, height = plotHeight)
# Plots all sets of points onto single graph, as well as 95% threshold line
plot(total_means, ylim=c(0,110), col=plotColors[1], pch=16,
     xlab="Number of Individuals", ylab="Percent Diversity Capture",
     main=paste0("QUAC, Microsatellites (Subset: 91 samples; ", num_reps, " Replicates)"))
points(v.com_means, col=plotColors[2], pch=16)
points(com_means, col=plotColors[3], pch=16)
points(lowfr_means, col=plotColors[4], pch=16)
points(rare_means, col=plotColors[5], pch=16)
legend(x=80, y=70.13276, inset = 0.05, legend = c("Total","Very common","Common","Low frequency", "Rare"),
       col=plotColors, pch = c(20,20,20), cex=1, pt.cex = 2, bty="n", y.intersp = 1)
# Lines for 95% threshold
abline(h=95, col="black", lty=3); abline(v=min_95_QUAC.MSAT_subset, col="black")
# Text for number of individuals to capture 95% threshold, with position based on variable
mtext(text=paste0("Minimum sampling size (95%) = ", min_95_QUAC.MSAT_subset), 
      side=1, line=-1.5, at=min_95_QUAC.MSAT_subset-15)
# Turn off plotting device
dev.off()

# SNP_subset
# De novo
# Call png command, to save resampling plots to disk
png(file = paste0(resamplingDataDir, "QUAC.SNP.DN.R0-R80.Subset.png"), width = plotWidth, height = plotHeight)
# Set plotting window to stack 2 graphs vertically
par(mfcol=c(2,1), oma=rep(0.2,4))
# R0
# Plots all sets of points onto single graph, as well as 95% threshold line
plot(total_means, ylim=c(0,110), col=plotColors[1], pch=16,
     xlab="Number of Individuals", ylab="Percent Diversity Capture",
     main=paste0("QUAC, SNPs: De novo, R0 (Subset: 91 samples; ", num_reps, " Replicates)"))
points(v.com_means, col=plotColors[2], pch=16)
points(com_means, col=plotColors[3], pch=16)
points(lowfr_means, col=plotColors[4], pch=16)
points(rare_means, col=plotColors[5], pch=16)
legend(x=83, y=70.13276, inset = 0.05, legend = c("Total","Very common","Common","Low frequency", "Rare"),
       col=plotColors, pch = c(20,20,20), cex=1, pt.cex = 2, bty="n", y.intersp = 1)
# Lines for 95% threshold
abline(h=95, col="black", lty=3); abline(v=min_95_QUAC.SNP.DN.R0_subset, col="black")
# Text for number of individuals to capture 95% threshold, with position based on variable
mtext(text=paste0("Minimum sampling size (95%) = ", min_95_QUAC.SNP.DN.R0_subset), 
      side=1, line=-1.5, at=min_95_QUAC.SNP.DN.R0_subset-10)
# R80
# Plots all sets of points onto single graph, as well as 95% threshold line
plot(total_means, ylim=c(0,110), col=plotColors[1], pch=16,
     xlab="Number of Individuals", ylab="Percent Diversity Capture",
     main=paste0("QUAC, SNPs: De novo, R80 (Subset: 91 samples; ", num_reps, " Replicates)"))
points(v.com_means, col=plotColors[2], pch=16)
points(com_means, col=plotColors[3], pch=16)
points(lowfr_means, col=plotColors[4], pch=16)
points(rare_means, col=plotColors[5], pch=16)
legend(x=83, y=70.13276, inset = 0.05, legend = c("Total","Very common","Common","Low frequency", "Rare"),
       col=plotColors, pch = c(20,20,20), cex=1, pt.cex = 2, bty="n", y.intersp = 1)
# Lines for 95% threshold
abline(h=95, col="black", lty=3); abline(v=min_95_QUAC.SNP.DN.R80_subset, col="black")
# Text for number of individuals to capture 95% threshold, with position based on variable
mtext(text=paste0("Minimum sampling size (95%) = ", min_95_QUAC.SNP.DN.R80_subset), 
      side=1, line=-1.5, at=min_95_QUAC.SNP.DN.R80_subset-10)
# Turn off plotting device
dev.off()

# Reference
# Call png command, to save resampling plots to disk
png(file = paste0(resamplingDataDir, "QUAC.SNP.REF.R0-R80.Subset.png"), width = plotWidth, height = plotHeight)
# Set plotting window to stack 2 graphs vertically
par(mfcol=c(2,1), oma=rep(0.2,4))
# R0
# Plots all sets of points onto single graph, as well as 95% threshold line
plot(total_means, ylim=c(0,110), col=plotColors[1], pch=16,
     xlab="Number of Individuals", ylab="Percent Diversity Capture",
     main=paste0("QUAC, SNPs: Reference, R0 (Subset: 91 samples; ", num_reps, " Replicates)"))
points(v.com_means, col=plotColors[2], pch=16)
points(com_means, col=plotColors[3], pch=16)
points(lowfr_means, col=plotColors[4], pch=16)
points(rare_means, col=plotColors[5], pch=16)
legend(x=83, y=70.13276, inset = 0.05, legend = c("Total","Very common","Common","Low frequency", "Rare"),
       col=plotColors, pch = c(20,20,20), cex=1, pt.cex = 2, bty="n", y.intersp = 1)
# Lines for 95% threshold
abline(h=95, col="black", lty=3); abline(v=min_95_QUAC.SNP.REF.R0_subset, col="black")
# Text for number of individuals to capture 95% threshold, with position based on variable
mtext(text=paste0("Minimum sampling size (95%) = ", min_95_QUAC.SNP.REF.R0_subset), 
      side=1, line=-1.5, at=min_95_QUAC.SNP.REF.R0_subset-10)
# R80
# Plots all sets of points onto single graph, as well as 95% threshold line
plot(total_means, ylim=c(0,110), col=plotColors[1], pch=16,
     xlab="Number of Individuals", ylab="Percent Diversity Capture",
     main=paste0("QUAC, SNPs: Reference, R80 (Subset: 91 samples; ", num_reps, " Replicates)"))
points(v.com_means, col=plotColors[2], pch=16)
points(com_means, col=plotColors[3], pch=16)
points(lowfr_means, col=plotColors[4], pch=16)
points(rare_means, col=plotColors[5], pch=16)
legend(x=83, y=70.13276, inset = 0.05, legend = c("Total","Very common","Common","Low frequency", "Rare"),
       col=plotColors, pch = c(20,20,20), cex=1, pt.cex = 2, bty="n", y.intersp = 1)
# Lines for 95% threshold
abline(h=95, col="black", lty=3); abline(v=min_95_QUAC.SNP.REF.R80_subset, col="black")
# Text for number of individuals to capture 95% threshold, with position based on variable
mtext(text=paste0("Minimum sampling size (95%) = ", min_95_QUAC.SNP.REF.R80_subset), 
      side=1, line=-1.5, at=min_95_QUAC.SNP.REF.R80_subset-10)
# Turn off plotting device
dev.off()

# %%%% QUBO %%%% ----
# ---- MSATS (COMPLETE) ----
# Read in genind file (Southeast Oaks repo; genetic_data/Qb_total.gen file)
genpop.filePath <- 
  "~/Documents/peripheralProjects/SE_oaks_genetics/genetic_data/"
QUBO.MSAT.genind <- read.genepop(paste0(genpop.filePath,"Qb_total.gen"), ncode=3)
# Correct popNames: last population (IMLS4_MP1_IMLS336_C05) is garden; rest (9) are wild
levels(QUBO.MSAT.genind@pop) <- c(rep("wild",9), "garden") 
# Rename samples: Split sample names on underscore, and return 3rd element. This is the "IMLS" portion
QUBO.MSAT.sampleNames <- unlist(lapply(rownames(QUBO.MSAT.genind@tab),function(x) strsplit(x, "_")[[1]][3]))
rownames(QUBO.MSAT.genind@tab) <- QUBO.MSAT.sampleNames

# CREATE RESAMPLING ARRAY AND CALCULATE SUMMARY STATISTICS
# Export genind object
clusterExport(cl, varlist = "QUBO.MSAT.genind")
# Run resampling in parallel, to generate an array that's saved to disc
samplingResults_QUBO.MSAT <- 
  exSitu_Resample_Parallel(QUBO.MSAT.genind, cluster = cl, reps = num_reps,
                           arrayFilepath=paste0(resamplingDataDir,"QUBO.MSAT.Complete_resampArr.Rdata"))
# Close cores
# stopCluster(cl)

# Average results across replicates (slices) of the sampling array, to determine
# the minimum number of samples required to capture 95% wild genetic diversity
# (We average samplingResults[,1,], since this column contains the total genetic diversity)
# Standard deviation is calculated as well
min_95_QUBO.MSAT <- min(which(apply(samplingResults_QUBO.MSAT[,1,],1,mean) > 95))
min_95.SD_QUBO.MSAT <- apply(samplingResults_QUBO.MSAT[,1,],1,sd)[min_95_QUBO.MSAT]
print(min_95_QUBO.MSAT); print(min_95.SD_QUBO.MSAT)
# Calculate means, for each capture rate category
total_means <- apply(samplingResults_QUBO.MSAT[,1,], 1, mean)
v.com_means <- apply(samplingResults_QUBO.MSAT[,2,], 1, mean)
com_means <- apply(samplingResults_QUBO.MSAT[,3,], 1, mean)
lowfr_means <- apply(samplingResults_QUBO.MSAT[,4,], 1, mean)
rare_means <- apply(samplingResults_QUBO.MSAT[,5,], 1, mean)

# *** PLOTTING ----
# Call png command, to save resampling plots to disk
png(file = paste0(resamplingDataDir, "QUBO.MSAT.Complete.png"), width = plotWidth, height = plotHeight)
# Plots all sets of points onto single graph, as well as 95% threshold line
plot(total_means, ylim=c(0,110), col=plotColors[1], pch=16,
     xlab="Number of Individuals", ylab="Percent Diversity Capture",
     main=paste0("QUBO, Microsatellites (Complete: 245 samples; ", num_reps, " Replicates)"))
points(v.com_means, col=plotColors[2], pch=16)
points(com_means, col=plotColors[3], pch=16)
points(lowfr_means, col=plotColors[4], pch=16)
points(rare_means, col=plotColors[5], pch=16)
legend(x=200, y=86.13276, inset = 0.05, legend = c("Total","Very common","Common","Low frequency", "Rare"),
       col=plotColors, pch = c(20,20,20), cex=1, pt.cex = 2, bty="n", y.intersp = 1)
# Lines for 95% threshold
abline(h=95, col="black", lty=3); abline(v=min_95_QUBO.MSAT, col="black")
# Text for number of individuals to capture 95% threshold, with position based on variable
mtext(text=paste0("Minimum sampling size (95%) = ", min_95_QUBO.MSAT), 
      side=1, line=-1.5, at=min_95_QUBO.MSAT-23)
# Turn off plotting device
dev.off()

# ---- SNPS (COMPLETE) ----
# %%% DE NOVO ----
# R0 ----
# READ IN GENIND FILE (Optimized de novo assembly; R0, min-maf=0, first SNP/locus, 2 populations (garden and wild))
genpop.filePath <- 
  "/RAID1/IMLS_GCCO/Analysis/Stacks/denovo_finalAssemblies/QUBO/output/populations_R0_NOMAF_1SNP_2Pops/"
QUBO.SNP.DN.R0.genind <- read.genepop(paste0(genpop.filePath,"populations.snps.gen"))
# Correct popNames
pop(QUBO.SNP.DN.R0.genind) <- 
  factor(read.table(paste0(genpop.filePath, "QUBO_popmap_GardenWild"), header=FALSE)[,2])
# Rename samples: get rid of "QUBO_G/W_" prefix. This will just leave the "IMLS" portion
QUBO.SNP.DN.R0.sampleNames <- gsub("QUBO_G_",replacement = "", row.names(QUBO.SNP.DN.R0.genind@tab))
QUBO.SNP.DN.R0.sampleNames <- gsub("QUBO_W_",replacement = "", QUBO.SNP.DN.R0.sampleNames)
# Replace SH-Q names in SNP list with IMLS names
# These were determined by Austin K., and are outlined on the Hoban Lab Drive ("MSATcomparisons_TissueNames")
# Only 1 of the 11 SH_Q garden samples has an IMLS sample name (SHQ2177); others are unshared 
QUBO.SNP.DN.R0.sampleNames <- gsub("SH_Q2177",replacement = "IMLS338", QUBO.SNP.DN.R0.sampleNames)
QUBO.SNP.DN.R0.sampleNames <- gsub("SH_Q2178",replacement = "IMLS312", QUBO.SNP.DN.R0.sampleNames)
QUBO.SNP.DN.R0.sampleNames <- gsub("SH_Q2179",replacement = "IMLS062", QUBO.SNP.DN.R0.sampleNames)
QUBO.SNP.DN.R0.sampleNames <- gsub("SH_Q2180",replacement = "IMLS051", QUBO.SNP.DN.R0.sampleNames)
QUBO.SNP.DN.R0.sampleNames <- gsub("SH_Q2181",replacement = "IMLS011", QUBO.SNP.DN.R0.sampleNames)
QUBO.SNP.DN.R0.sampleNames <- gsub("SH_Q2182",replacement = "IMLS144", QUBO.SNP.DN.R0.sampleNames)
QUBO.SNP.DN.R0.sampleNames <- gsub("SH_Q2183",replacement = "IMLS170", QUBO.SNP.DN.R0.sampleNames)
QUBO.SNP.DN.R0.sampleNames <- gsub("SH_Q2184",replacement = "IMLS005", QUBO.SNP.DN.R0.sampleNames)
QUBO.SNP.DN.R0.sampleNames <- gsub("SH_Q2186",replacement = "IMLS017", QUBO.SNP.DN.R0.sampleNames)
# Rename sample matrix
rownames(QUBO.SNP.DN.R0.genind@tab) <- QUBO.SNP.DN.R0.sampleNames

# CREATE RESAMPLING ARRAY AND CALCULATE SUMMARY STATISTICS
# Export genind object
clusterExport(cl, varlist = "QUBO.SNP.DN.R0.genind")
# Run resampling in parallel, to generate an array that's saved to disc
samplingResults_QUBO.SNP.DN.R0 <- 
  exSitu_Resample_Parallel(QUBO.SNP.DN.R0.genind, cluster = cl, reps = num_reps,
                           arrayFilepath=paste0(resamplingDataDir,"QUBO.SNP.DN.R0.Complete_resampArr.Rdata"))
# Close cores
# stopCluster(cl)

# Average results across replicates (slices) of the sampling array, to determine
# the minimum number of samples required to capture 95% wild genetic diversity
# (We average samplingResults[,1,], since this column contains the total genetic diversity)
# Standard deviation is calculated as well
min_95_QUBO.SNP.DN.R0 <- min(which(apply(samplingResults_QUBO.SNP.DN.R0[,1,],1,mean) > 95))
min_95.SD_QUBO.SNP.DN.R0 <- apply(samplingResults_QUBO.SNP.DN.R0[,1,],1,sd)[min_95_QUBO.SNP.DN.R0]
print(min_95_QUBO.SNP.DN.R0); print(min_95.SD_QUBO.SNP.DN.R0)
# Calculate means, for each capture rate category
total_means <- apply(samplingResults_QUBO.SNP.DN.R0[,1,], 1, mean)
v.com_means <- apply(samplingResults_QUBO.SNP.DN.R0[,2,], 1, mean)
com_means <- apply(samplingResults_QUBO.SNP.DN.R0[,3,], 1, mean)
lowfr_means <- apply(samplingResults_QUBO.SNP.DN.R0[,4,], 1, mean)
rare_means <- apply(samplingResults_QUBO.SNP.DN.R0[,5,], 1, mean)

# R80 ----
# READ IN GENIND FILE (Optimized de novo assembly; R80, min-maf=0, first SNP/locus, 2 populations (garden and wild))
genpop.filePath <- 
  "/RAID1/IMLS_GCCO/Analysis/Stacks/denovo_finalAssemblies/QUBO/output/populations_R80_NOMAF_1SNP_2Pops/"
QUBO.SNP.DN.R80.genind <- read.genepop(paste0(genpop.filePath,"populations.snps.gen"))
# Correct popNames
pop(QUBO.SNP.DN.R80.genind) <- 
  factor(read.table(paste0(genpop.filePath, "QUBO_popmap_GardenWild"), header=FALSE)[,2])
# Rename sample matrix, based on QUBO.SNP.DN.R0.sampleNames
rownames(QUBO.SNP.DN.R80.genind@tab) <- QUBO.SNP.DN.R0.sampleNames

# CREATE RESAMPLING ARRAY AND CALCULATE SUMMARY STATISTICS
# Export genind object
clusterExport(cl, varlist = "QUBO.SNP.DN.R80.genind")
# Run resampling in parallel, to generate an array that's saved to disc
samplingResults_QUBO.SNP.DN.R80 <- 
  exSitu_Resample_Parallel(QUBO.SNP.DN.R80.genind, cluster = cl, reps = num_reps,
                           arrayFilepath=paste0(resamplingDataDir,"QUBO.SNP.DN.R80.Complete_resampArr.Rdata"))
# Close cores
# stopCluster(cl)

# Average results across replicates (slices) of the sampling array, to determine
# the minimum number of samples required to capture 95% wild genetic diversity
# (We average samplingResults[,1,], since this column contains the total genetic diversity)
# Standard deviation is calculated as well
min_95_QUBO.SNP.DN.R80 <- min(which(apply(samplingResults_QUBO.SNP.DN.R80[,1,],1,mean) > 95))
min_95.SD_QUBO.SNP.DN.R80 <- apply(samplingResults_QUBO.SNP.DN.R80[,1,],1,sd)[min_95_QUBO.SNP.DN.R80]
print(min_95_QUBO.SNP.DN.R80); print(min_95.SD_QUBO.SNP.DN.R80)
# Calculate means, for each capture rate category
total_means <- apply(samplingResults_QUBO.SNP.DN.R80[,1,], 1, mean)
v.com_means <- apply(samplingResults_QUBO.SNP.DN.R80[,2,], 1, mean)
com_means <- apply(samplingResults_QUBO.SNP.DN.R80[,3,], 1, mean)
lowfr_means <- apply(samplingResults_QUBO.SNP.DN.R80[,4,], 1, mean)
rare_means <- apply(samplingResults_QUBO.SNP.DN.R80[,5,], 1, mean)

# *** PLOTTING ----
# Call png command, to save resampling plots to disk
png(file = paste0(resamplingDataDir, "QUBO.SNP.DN.R0-R80.Complete.png"), width = plotWidth, height = plotHeight)
# Set plotting window to stack 2 graphs vertically
par(mfcol=c(2,1), oma=rep(0.2,4))
# R0
# Plots all sets of points onto single graph, as well as 95% threshold line
plot(total_means, ylim=c(0,110), col=plotColors[1], pch=16,
     xlab="Number of Individuals", ylab="Percent Diversity Capture",
     main=paste0("QUBO, SNPs: De novo, R0 (Complete: 95 samples; ", num_reps, " Replicates)"))
points(v.com_means, col=plotColors[2], pch=16)
points(com_means, col=plotColors[3], pch=16)
points(lowfr_means, col=plotColors[4], pch=16)
points(rare_means, col=plotColors[5], pch=16)
legend(x=83, y=86.13276, inset = 0.05, legend = c("Total","Very common","Common","Low frequency", "Rare"),
       col=plotColors, pch = c(20,20,20), cex=1, pt.cex = 2, bty="n", y.intersp = 1)
# Lines for 95% threshold
abline(h=95, col="black", lty=3); abline(v=min_95_QUBO.SNP.DN.R0, col="black")
# Text for number of individuals to capture 95% threshold, with position based on variable
mtext(text=paste0("Minimum sampling size (95%) = ", min_95_QUBO.SNP.DN.R0), 
      side=1, line=-1.5, at=min_95_QUBO.SNP.DN.R0-10)
# R80
# Plots all sets of points onto single graph, as well as 95% threshold line
plot(total_means, ylim=c(0,110), col=plotColors[1], pch=16,
     xlab="Number of Individuals", ylab="Percent Diversity Capture",
     main=paste0("QUBO, SNPs: De novo, R80 (Complete: 95 samples; ", num_reps, " Replicates)"))
points(v.com_means, col=plotColors[2], pch=16)
points(com_means, col=plotColors[3], pch=16)
points(lowfr_means, col=plotColors[4], pch=16)
points(rare_means, col=plotColors[5], pch=16)
legend(x=83, y=86.13276, inset = 0.05, legend = c("Total","Very common","Common","Low frequency", "Rare"),
       col=plotColors, pch = c(20,20,20), cex=1, pt.cex = 2, bty="n", y.intersp = 1)
# Lines for 95% threshold
abline(h=95, col="black", lty=3); abline(v=min_95_QUBO.SNP.DN.R80, col="black")
# Text for number of individuals to capture 95% threshold, with position based on variable
mtext(text=paste0("Minimum sampling size (95%) = ", min_95_QUBO.SNP.DN.R80), 
      side=1, line=-1.5, at=min_95_QUBO.SNP.DN.R80-10)
# Turn off plotting device
dev.off()

# %%% REFERENCE ----
# R0 ----
# READ IN GENIND FILE (QUBO GSNAP4 alignment; R0, min-maf=0, first SNP/locus, 2 populations (garden and wild))
genpop.filePath <- 
  "/RAID1/IMLS_GCCO/Analysis/Stacks/reference_filteredReads/QUBO/GSNAP4/output/populations_R0_NOMAF_1SNP_2Pops/"
QUBO.SNP.REF.R0.genind <- read.genepop(paste0(genpop.filePath,"populations.snps.gen"))
# Correct popNames
pop(QUBO.SNP.REF.R0.genind) <- 
  factor(read.table(paste0(genpop.filePath, "QUBO_popmap_GardenWild"), header=FALSE)[,2])
# Rename sample matrix, based on QUBO.SNP.DN.R0.sampleNames
rownames(QUBO.SNP.REF.R0.genind@tab) <- QUBO.SNP.DN.R0.sampleNames

# CREATE RESAMPLING ARRAY AND CALCULATE SUMMARY STATISTICS
# Export genind object
clusterExport(cl, varlist = "QUBO.SNP.REF.R0.genind")
# Run resampling in parallel, to generate an array that's saved to disc
samplingResults_QUBO.SNP.REF.R0 <- 
  exSitu_Resample_Parallel(QUBO.SNP.REF.R0.genind, cluster = cl, reps = num_reps,
                           arrayFilepath=paste0(resamplingDataDir,"QUBO.SNP.REF.R0.Complete_resampArr.Rdata"))
# Close cores
# stopCluster(cl)

# Average results across replicates (slices) of the sampling array, to determine
# the minimum number of samples required to capture 95% wild genetic diversity
# (We average samplingResults[,1,], since this column contains the total genetic diversity)
# Standard deviation is calculated as well
min_95_QUBO.SNP.REF.R0 <- min(which(apply(samplingResults_QUBO.SNP.REF.R0[,1,],1,mean) > 95))
min_95.SD_QUBO.SNP.REF.R0 <- apply(samplingResults_QUBO.SNP.REF.R0[,1,],1,sd)[min_95_QUBO.SNP.REF.R0]
print(min_95_QUBO.SNP.REF.R0); print(min_95.SD_QUBO.SNP.REF.R0)
# Calculate means, for each capture rate category
total_means <- apply(samplingResults_QUBO.SNP.REF.R0[,1,], 1, mean)
v.com_means <- apply(samplingResults_QUBO.SNP.REF.R0[,2,], 1, mean)
com_means <- apply(samplingResults_QUBO.SNP.REF.R0[,3,], 1, mean)
lowfr_means <- apply(samplingResults_QUBO.SNP.REF.R0[,4,], 1, mean)
rare_means <- apply(samplingResults_QUBO.SNP.REF.R0[,5,], 1, mean)

# R80 ----
# READ IN GENIND FILE (QUBO GSNAP4 alignment; R80, min-maf=0, first SNP/locus, 2 populations (garden and wild))
genpop.filePath <- 
  "/RAID1/IMLS_GCCO/Analysis/Stacks/reference_filteredReads/QUBO/GSNAP4/output/populations_R80_NOMAF_1SNP_2Pops/"
QUBO.SNP.REF.R80.genind <- read.genepop(paste0(genpop.filePath,"populations.snps.gen"))
# Correct popNames
pop(QUBO.SNP.REF.R80.genind) <- 
  factor(read.table(paste0(genpop.filePath, "QUBO_popmap_GardenWild"), header=FALSE)[,2])
# Rename sample matrix, based on QUBO.SNP.DN.R0.sampleNames
rownames(QUBO.SNP.REF.R80.genind@tab) <- QUBO.SNP.DN.R0.sampleNames

# CREATE RESAMPLING ARRAY AND CALCULATE SUMMARY STATISTICS
# Export genind object
clusterExport(cl, varlist = "QUBO.SNP.REF.R80.genind")
# Run resampling in parallel, to generate an array that's saved to disc
samplingResults_QUBO.SNP.REF.R80 <- 
  exSitu_Resample_Parallel(QUBO.SNP.REF.R80.genind, cluster = cl, reps = num_reps,
                           arrayFilepath=paste0(resamplingDataDir,"QUBO.SNP.REF.R80.Complete_resampArr.Rdata"))
# Close cores
# stopCluster(cl)

# Average results across replicates (slices) of the sampling array, to determine
# the minimum number of samples required to capture 95% wild genetic diversity
# (We average samplingResults[,1,], since this column contains the total genetic diversity)
# Standard deviation is calculated as well
min_95_QUBO.SNP.REF.R80 <- min(which(apply(samplingResults_QUBO.SNP.REF.R80[,1,],1,mean) > 95))
min_95.SD_QUBO.SNP.REF.R80 <- apply(samplingResults_QUBO.SNP.REF.R80[,1,],1,sd)[min_95_QUBO.SNP.REF.R80]
print(min_95_QUBO.SNP.REF.R80); print(min_95.SD_QUBO.SNP.REF.R80)
# Calculate means, for each capture rate category
total_means <- apply(samplingResults_QUBO.SNP.REF.R80[,1,], 1, mean)
v.com_means <- apply(samplingResults_QUBO.SNP.REF.R80[,2,], 1, mean)
com_means <- apply(samplingResults_QUBO.SNP.REF.R80[,3,], 1, mean)
lowfr_means <- apply(samplingResults_QUBO.SNP.REF.R80[,4,], 1, mean)
rare_means <- apply(samplingResults_QUBO.SNP.REF.R80[,5,], 1, mean)

# *** PLOTTING ----
# Call png command, to save resampling plots to disk
png(file = paste0(resamplingDataDir, "QUBO.SNP.REF.R0-R80.Complete.png"), width = plotWidth, height = plotHeight)
# Set plotting window to stack 2 graphs vertically
par(mfcol=c(2,1), oma=rep(0.2,4))
# R0
# Plots all sets of points onto single graph, as well as 95% threshold line
plot(total_means, ylim=c(0,110), col=plotColors[1], pch=16,
     xlab="Number of Individuals", ylab="Percent Diversity Capture",
     main=paste0("QUBO, SNPs: Reference, R0 (Complete: 95 samples; ", num_reps, " Replicates)"))
points(v.com_means, col=plotColors[2], pch=16)
points(com_means, col=plotColors[3], pch=16)
points(lowfr_means, col=plotColors[4], pch=16)
points(rare_means, col=plotColors[5], pch=16)
legend(x=83, y=86.13276, inset = 0.05, legend = c("Total","Very common","Common","Low frequency", "Rare"),
       col=plotColors, pch = c(20,20,20), cex=1, pt.cex = 2, bty="n", y.intersp = 1)
# Lines for 95% threshold
abline(h=95, col="black", lty=3); abline(v=min_95_QUBO.SNP.REF.R0, col="black")
# Text for number of individuals to capture 95% threshold, with position based on variable
mtext(text=paste0("Minimum sampling size (95%) = ", min_95_QUBO.SNP.REF.R0), 
      side=1, line=-1.5, at=min_95_QUBO.SNP.REF.R0-10)
# R80
# Plots all sets of points onto single graph, as well as 95% threshold line
plot(total_means, ylim=c(0,110), col=plotColors[1], pch=16,
     xlab="Number of Individuals", ylab="Percent Diversity Capture",
     main=paste0("QUBO, SNPs: Reference, R80 (Complete: 95 samples; ", num_reps, " Replicates)"))
points(v.com_means, col=plotColors[2], pch=16)
points(com_means, col=plotColors[3], pch=16)
points(lowfr_means, col=plotColors[4], pch=16)
points(rare_means, col=plotColors[5], pch=16)
legend(x=83, y=86.13276, inset = 0.05, legend = c("Total","Very common","Common","Low frequency", "Rare"),
       col=plotColors, pch = c(20,20,20), cex=1, pt.cex = 2, bty="n", y.intersp = 1)
# Lines for 95% threshold
abline(h=95, col="black", lty=3); abline(v=min_95_QUBO.SNP.REF.R80, col="black")
# Text for number of individuals to capture 95% threshold, with position based on variable
mtext(text=paste0("Minimum sampling size (95%) = ", min_95_QUBO.SNP.REF.R80), 
      side=1, line=-1.5, at=min_95_QUBO.SNP.REF.R80-10)
# Turn off plotting device
dev.off()

# ---- MSATS AND SNPS: SUBSET ----
# Subset SNP sample names by those that are also seen within the MSAT samples
QUBO_sharedSamples <- sort(QUBO.SNP.DN.R0.sampleNames[which(QUBO.SNP.DN.R0.sampleNames %in% QUBO.MSAT.sampleNames)])
# Subset MSAT and SNP wild matrix objects to strictly shared samples
QUBO.MSAT_subset.genind <- QUBO.MSAT.genind[QUBO_sharedSamples,, drop=TRUE]
# De novo
QUBO.SNP.DN.R0_subset.genind <- QUBO.SNP.DN.R0.genind[QUBO_sharedSamples,, drop=TRUE]
QUBO.SNP.DN.R80_subset.genind <- QUBO.SNP.DN.R80.genind[QUBO_sharedSamples,, drop=TRUE]
# Reference
QUBO.SNP.REF.R0_subset.genind <- QUBO.SNP.REF.R0.genind[QUBO_sharedSamples,, drop=TRUE]
QUBO.SNP.REF.R80_subset.genind <- QUBO.SNP.REF.R80.genind[QUBO_sharedSamples,, drop=TRUE]

# CREATE RESAMPLING ARRAY AND CALCULATE SUMMARY STATISTICS
# Export genind objects
clusterExport(cl, varlist = c("QUBO.MSAT_subset.genind", "QUBO.SNP.DN.R0_subset.genind",
                              "QUBO.SNP.DN.R80_subset.genind", "QUBO.SNP.REF.R0_subset.genind",
                              "QUBO.SNP.REF.R80_subset.genind"))
# Run resampling in parallel, to generate arrays
# MSAT
samplingResults_QUBO.MSAT_subset <- 
  exSitu_Resample_Parallel(QUBO.MSAT_subset.genind, cluster = cl, reps = num_reps,
                           arrayFilepath=paste0(resamplingDataDir,"QUBO.MSAT.Subset_resampArr.Rdata"))
# SNP: De novo
# R0
samplingResults_QUBO.SNP.DN.R0_subset <- 
  exSitu_Resample_Parallel(QUBO.SNP.DN.R0_subset.genind, cluster = cl, reps = num_reps,
                           arrayFilepath=paste0(resamplingDataDir,"QUBO.SNP.DN.R0.Subset_resampArr.Rdata"))
# R80
samplingResults_QUBO.SNP.DN.R80_subset <- 
  exSitu_Resample_Parallel(QUBO.SNP.DN.R80_subset.genind, cluster = cl, reps = num_reps,
                           arrayFilepath=paste0(resamplingDataDir,"QUBO.SNP.DN.R80.Subset_resampArr.Rdata"))
# SNP: Reference
# R0
samplingResults_QUBO.SNP.REF.R0_subset <- 
  exSitu_Resample_Parallel(QUBO.SNP.REF.R0_subset.genind, cluster = cl, reps = num_reps,
                           arrayFilepath=paste0(resamplingDataDir,"QUBO.SNP.REF.R0.Subset_resampArr.Rdata"))
# R80
samplingResults_QUBO.SNP.REF.R80_subset <- 
  exSitu_Resample_Parallel(QUBO.SNP.REF.R80_subset.genind, cluster = cl, reps = num_reps,
                           arrayFilepath=paste0(resamplingDataDir,"QUBO.SNP.REF.R80.Subset_resampArr.Rdata"))
# Close cores
stopCluster(cl)

# MSAT_subset: average 95% minimum sampling size across replicates, as well as standard deviation
min_95_QUBO.MSAT_subset <- min(which(apply(samplingResults_QUBO.MSAT_subset[,1,],1,mean) > 95))
min_95.SD_QUBO.MSAT_subset <- apply(samplingResults_QUBO.MSAT_subset[,1,],1,sd)[min_95_QUBO.MSAT_subset]
print(min_95_QUBO.MSAT_subset); print(min_95.SD_QUBO.MSAT_subset)
# MSAT_subset: calculate means and standard deviations, for each capture rate category
total_means <- apply(samplingResults_QUBO.MSAT_subset[,1,], 1, mean)
v.com_means <- apply(samplingResults_QUBO.MSAT_subset[,2,], 1, mean)
com_means <- apply(samplingResults_QUBO.MSAT_subset[,3,], 1, mean)
lowfr_means <- apply(samplingResults_QUBO.MSAT_subset[,4,], 1, mean)
rare_means <- apply(samplingResults_QUBO.MSAT_subset[,5,], 1, mean)

# SNP_subset: average 95% minimum sampling size across replicates, as well as standard deviation
# De novo 
# R0
min_95_QUBO.SNP.DN.R0_subset <- min(which(apply(samplingResults_QUBO.SNP.DN.R0_subset[,1,],1,mean) > 95))
min_95.SD_QUBO.SNP.DN.R0_subset <- apply(samplingResults_QUBO.SNP.DN.R0_subset[,1,],1,sd)[min_95_QUBO.SNP.DN.R0_subset]
print(min_95_QUBO.SNP.DN.R0_subset); print(min_95.SD_QUBO.SNP.DN.R0_subset)
# Calculate means, for each capture rate category
total_means <- apply(samplingResults_QUBO.SNP.DN.R0_subset[,1,], 1, mean)
v.com_means <- apply(samplingResults_QUBO.SNP.DN.R0_subset[,2,], 1, mean)
com_means <- apply(samplingResults_QUBO.SNP.DN.R0_subset[,3,], 1, mean)
lowfr_means <- apply(samplingResults_QUBO.SNP.DN.R0_subset[,4,], 1, mean)
rare_means <- apply(samplingResults_QUBO.SNP.DN.R0_subset[,5,], 1, mean)
# R80
min_95_QUBO.SNP.DN.R80_subset <- min(which(apply(samplingResults_QUBO.SNP.DN.R80_subset[,1,],1,mean) > 95))
min_95.SD_QUBO.SNP.DN.R80_subset <- apply(samplingResults_QUBO.SNP.DN.R80_subset[,1,],1,sd)[min_95_QUBO.SNP.DN.R80_subset]
print(min_95_QUBO.SNP.DN.R80_subset); print(min_95.SD_QUBO.SNP.DN.R80_subset)
# Calculate means, for each capture rate category
total_means <- apply(samplingResults_QUBO.SNP.DN.R80_subset[,1,], 1, mean)
v.com_means <- apply(samplingResults_QUBO.SNP.DN.R80_subset[,2,], 1, mean)
com_means <- apply(samplingResults_QUBO.SNP.DN.R80_subset[,3,], 1, mean)
lowfr_means <- apply(samplingResults_QUBO.SNP.DN.R80_subset[,4,], 1, mean)
rare_means <- apply(samplingResults_QUBO.SNP.DN.R80_subset[,5,], 1, mean)

# Reference
# R0
min_95_QUBO.SNP.REF.R0_subset <- min(which(apply(samplingResults_QUBO.SNP.REF.R0_subset[,1,],1,mean) > 95))
min_95.SD_QUBO.SNP.REF.R0_subset <- apply(samplingResults_QUBO.SNP.REF.R0_subset[,1,],1,sd)[min_95_QUBO.SNP.REF.R0_subset]
print(min_95_QUBO.SNP.REF.R0_subset); print(min_95.SD_QUBO.SNP.REF.R0_subset)
# Calculate means, for each capture rate category
total_means <- apply(samplingResults_QUBO.SNP.REF.R0_subset[,1,], 1, mean)
v.com_means <- apply(samplingResults_QUBO.SNP.REF.R0_subset[,2,], 1, mean)
com_means <- apply(samplingResults_QUBO.SNP.REF.R0_subset[,3,], 1, mean)
lowfr_means <- apply(samplingResults_QUBO.SNP.REF.R0_subset[,4,], 1, mean)
rare_means <- apply(samplingResults_QUBO.SNP.REF.R0_subset[,5,], 1, mean)
# R80
min_95_QUBO.SNP.REF.R80_subset <- min(which(apply(samplingResults_QUBO.SNP.REF.R80_subset[,1,],1,mean) > 95))
min_95.SD_QUBO.SNP.REF.R80_subset <- apply(samplingResults_QUBO.SNP.REF.R80_subset[,1,],1,sd)[min_95_QUBO.SNP.REF.R80_subset]
print(min_95_QUBO.SNP.REF.R80_subset); print(min_95.SD_QUBO.SNP.REF.R80_subset)
# Calculate means, for each capture rate category
total_means <- apply(samplingResults_QUBO.SNP.REF.R80_subset[,1,], 1, mean)
v.com_means <- apply(samplingResults_QUBO.SNP.REF.R80_subset[,2,], 1, mean)
com_means <- apply(samplingResults_QUBO.SNP.REF.R80_subset[,3,], 1, mean)
lowfr_means <- apply(samplingResults_QUBO.SNP.REF.R80_subset[,4,], 1, mean)
rare_means <- apply(samplingResults_QUBO.SNP.REF.R80_subset[,5,], 1, mean)

# *** PLOTTING ----
# Call png command, to save resampling plots to disk
png(file = paste0(resamplingDataDir, "QUBO.MSAT.Subset.png"), width = plotWidth, height = plotHeight)
# Plots all sets of points onto single graph, as well as 95% threshold line
plot(total_means, ylim=c(0,110), col=plotColors[1], pch=16,
     xlab="Number of Individuals", ylab="Percent Diversity Capture",
     main=paste0("QUBO, Microsatellites (Subset: 94 samples; ", num_reps, " Replicates)"))
points(v.com_means, col=plotColors[2], pch=16)
points(com_means, col=plotColors[3], pch=16)
points(lowfr_means, col=plotColors[4], pch=16)
points(rare_means, col=plotColors[5], pch=16)
legend(x=80, y=70.13276, inset = 0.05, legend = c("Total","Very common","Common","Low frequency", "Rare"),
       col=plotColors, pch = c(20,20,20), cex=1, pt.cex = 2, bty="n", y.intersp = 1)
# Lines for 95% threshold
abline(h=95, col="black", lty=3); abline(v=min_95_QUBO.MSAT_subset, col="black")
# Text for number of individuals to capture 95% threshold, with position based on variable
mtext(text=paste0("Minimum sampling size (95%) = ", min_95_QUBO.MSAT_subset), 
      side=1, line=-1.5, at=min_95_QUBO.MSAT_subset-10)
# Turn off plotting device
dev.off()

# SNP_subset
# De novo
# Call png command, to save resampling plots to disk
png(file = paste0(resamplingDataDir, "QUBO.SNP.DN.R0-R80.Subset.png"), width = plotWidth, height = plotHeight)
# Set plotting window to stack 2 graphs vertically
par(mfcol=c(2,1), oma=rep(0.2,4))
# R0
# Plots all sets of points onto single graph, as well as 95% threshold line
plot(total_means, ylim=c(0,110), col=plotColors[1], pch=16,
     xlab="Number of Individuals", ylab="Percent Diversity Capture",
     main=paste0("QUBO, SNPs: De novo, R0 (Subset: 94 samples; ", num_reps, " Replicates)"))
points(v.com_means, col=plotColors[2], pch=16)
points(com_means, col=plotColors[3], pch=16)
points(lowfr_means, col=plotColors[4], pch=16)
points(rare_means, col=plotColors[5], pch=16)
legend(x=83, y=70.13276, inset = 0.05, legend = c("Total","Very common","Common","Low frequency", "Rare"),
       col=plotColors, pch = c(20,20,20), cex=1, pt.cex = 2, bty="n", y.intersp = 1)
# Lines for 95% threshold
abline(h=95, col="black", lty=3); abline(v=min_95_QUBO.SNP.DN.R0_subset, col="black")
# Text for number of individuals to capture 95% threshold, with position based on variable
mtext(text=paste0("Minimum sampling size (95%) = ", min_95_QUBO.SNP.DN.R0_subset), 
      side=1, line=-1.5, at=min_95_QUBO.SNP.DN.R0_subset-10)
# R80
# Plots all sets of points onto single graph, as well as 95% threshold line
plot(total_means, ylim=c(0,110), col=plotColors[1], pch=16,
     xlab="Number of Individuals", ylab="Percent Diversity Capture",
     main=paste0("QUBO, SNPs: De novo, R80 (Subset: 94 samples; ", num_reps, " Replicates)"))
points(v.com_means, col=plotColors[2], pch=16)
points(com_means, col=plotColors[3], pch=16)
points(lowfr_means, col=plotColors[4], pch=16)
points(rare_means, col=plotColors[5], pch=16)
legend(x=83, y=70.13276, inset = 0.05, legend = c("Total","Very common","Common","Low frequency", "Rare"),
       col=plotColors, pch = c(20,20,20), cex=1, pt.cex = 2, bty="n", y.intersp = 1)
# Lines for 95% threshold
abline(h=95, col="black", lty=3); abline(v=min_95_QUBO.SNP.DN.R80_subset, col="black")
# Text for number of individuals to capture 95% threshold, with position based on variable
mtext(text=paste0("Minimum sampling size (95%) = ", min_95_QUBO.SNP.DN.R80_subset), 
      side=1, line=-1.5, at=min_95_QUBO.SNP.DN.R80_subset-10)
# Turn off plotting device
dev.off()

# Reference
# Call png command, to save resampling plots to disk
png(file = paste0(resamplingDataDir, "QUBO.SNP.REF.R0-R80.Subset.png"), width = plotWidth, height = plotHeight)
# Set plotting window to stack 2 graphs vertically
par(mfcol=c(2,1), oma=rep(0.2,4))
# R0
# Plots all sets of points onto single graph, as well as 95% threshold line
plot(total_means, ylim=c(0,110), col=plotColors[1], pch=16,
     xlab="Number of Individuals", ylab="Percent Diversity Capture",
     main=paste0("QUBO, SNPs: Reference, R0 (Subset: 94 samples; ", num_reps, " Replicates)"))
points(v.com_means, col=plotColors[2], pch=16)
points(com_means, col=plotColors[3], pch=16)
points(lowfr_means, col=plotColors[4], pch=16)
points(rare_means, col=plotColors[5], pch=16)
legend(x=83, y=70.13276, inset = 0.05, legend = c("Total","Very common","Common","Low frequency", "Rare"),
       col=plotColors, pch = c(20,20,20), cex=1, pt.cex = 2, bty="n", y.intersp = 1)
# Lines for 95% threshold
abline(h=95, col="black", lty=3); abline(v=min_95_QUBO.SNP.REF.R0_subset, col="black")
# Text for number of individuals to capture 95% threshold, with position based on variable
mtext(text=paste0("Minimum sampling size (95%) = ", min_95_QUBO.SNP.REF.R0_subset), 
      side=1, line=-1.5, at=min_95_QUBO.SNP.REF.R0_subset-10)
# R80
# Plots all sets of points onto single graph, as well as 95% threshold line
plot(total_means, ylim=c(0,110), col=plotColors[1], pch=16,
     xlab="Number of Individuals", ylab="Percent Diversity Capture",
     main=paste0("QUBO, SNPs: Reference, R80 (Subset: 94 samples; ", num_reps, " Replicates)"))
points(v.com_means, col=plotColors[2], pch=16)
points(com_means, col=plotColors[3], pch=16)
points(lowfr_means, col=plotColors[4], pch=16)
points(rare_means, col=plotColors[5], pch=16)
legend(x=83, y=70.13276, inset = 0.05, legend = c("Total","Very common","Common","Low frequency", "Rare"),
       col=plotColors, pch = c(20,20,20), cex=1, pt.cex = 2, bty="n", y.intersp = 1)
# Lines for 95% threshold
abline(h=95, col="black", lty=3); abline(v=min_95_QUBO.SNP.REF.R80_subset, col="black")
# Text for number of individuals to capture 95% threshold, with position based on variable
mtext(text=paste0("Minimum sampling size (95%) = ", min_95_QUBO.SNP.REF.R80_subset), 
      side=1, line=-1.5, at=min_95_QUBO.SNP.REF.R80_subset-10)
# Turn off plotting device
dev.off()
