# %%%%%%%%%%%%%%%%%%
# %%% RESAMPLING %%%
# %%%%%%%%%%%%%%%%%%

# This script generates Resampling arrays, and plots their results, 
# for Quercus acerifolia (QUAC; optimized Stacks de novo assembly, m 7, M/n 4, gt-alpha 0.01) 
# and Quercus boyntonii (QUBO; GSNAP4 alignment with Quercus robur reference) NextRAD samples.

# Resampling arrays are collections of values indicating the percentage of total wild alleles
# that are represented by a (random) subset of wild samples. Each column of these arrays corresponds
# to the category of allelic representation of the random subset of wild samples (total allelic diversity,
# very common alleles, common alleles, low frequency alleles, and rare alleles).

# In addition to processing SNP datasets (described above), this script also 
# reads in the QUAC and QUBO microsatellite (MSAT) genind files as well.
# It analyzes both the complete SNP and MSAT datasets AND each dataset subset to the samples
# shared between the SNP and MSAT studies (for a more direct comparison).

# For QUAC, SNP and MSAT subsets are generated by using the Tissue database names for each SNP/MSAT sample
# For QUBO, SNP and MSAT subsets are generated by shortening names to only IMLS sample names

library(adegenet)
library(RColorBrewer)
library(scales)
library(parallel)

# %%%% FUNCTIONS AND PARALLELIZATION %%%% ----
# Read in relevant functions required for resampling analyses
SSRvSNP.wd <- "~/Documents/SSRvSNP/Code/"
setwd(SSRvSNP.wd)
source("exSituRepresentation/functions_exSituRepresentation.R")
# Set up relevant cores, and make sure adegenet library is present on cluster
num_cores <- detectCores() - 1 ; cl <- makeCluster(num_cores)
clusterEvalQ(cl, library("adegenet"))

# %%%% QUAC %%%% ----
# ---- MSATS (COMPLETE) ----
# READ IN GENIND FILE (GCC_QUAC_ZAIN repo; QUAC_wK_garden_wild_clean.gen) ----
genpop.filePath <- 
  "~/Documents/peripheralProjects/GCC_QUAC_ZAIN/Data_Files/Adegenet_Files/Garden_Wild/"
setwd(genpop.filePath)
QUAC.MSAT.genind <- read.genepop("QUAC_wK_garden_wild_clean.gen", ncode = 3)
# Correct popNames: pop1 is Garden, pop2 is Wild
pop(QUAC.MSAT.genind) <- gsub("pop1", "garden", pop(QUAC.MSAT.genind))
pop(QUAC.MSAT.genind) <- gsub("pop2", "wild", pop(QUAC.MSAT.genind))
# Correct sample names: read tissue database names (from GCC_QUAC_ZAIN repository), rename genind rows
QUAC.MSAT.tissueNames_filepath <- "~/Documents/peripheralProjects/GCC_QUAC_ZAIN/Data_Files/Data_Frames/QUAC_allpop_clean_df.csv"
QUAC.MSAT.tissueNames <- unlist(read.csv2(QUAC.MSAT.tissueNames_filepath, header = TRUE, sep=",")[1])
rownames(QUAC.MSAT.genind@tab) <- QUAC.MSAT.tissueNames

# CREATE SAMPLING RESULTS ARRAY ----
# Specify number of replicates
num_reps <- 50
# Export relevant functions and variables
clusterExport(cl, varlist = c("getAlleleCategories", "exSitu_Sample", "exSitu_Resample", 
                              "num_reps", "QUAC.MSAT.genind"))
# Run resampling in parallel, to generate an array
samplingResults_QUAC.MSAT <- 
  parSapply(cl, 1:num_reps, 
            function(a) exSitu_Resample(gen.obj = QUAC.MSAT.genind), simplify = "array")
# Close cores
# stopCluster(cl)

# Average results across replicates (slices) of the sampling array, to determine
# the minimum number of samples required to capture 95% wild genetic diversity
# (We average samplingResults[,1,], since this column contains the total genetic diversity)
# Standard deviation is calculated as well
min_95_QUAC.MSAT <- min(which(apply(samplingResults_QUAC.MSAT[,1,],1,mean) > 95))
min_95.SD_QUAC.MSAT <- apply(samplingResults_QUAC.MSAT[,1,],1,sd)[min_95_QUAC.MSAT]
print(min_95_QUAC.MSAT); print(min_95.SD_QUAC.MSAT)

# PLOTTING ----
# Set plotting window to stack 2 graphs vertically
par(mfcol=c(2,1), oma=rep(0.2,4))
# Calculate means and standard deviations, for each capture rate category
total_means <- apply(samplingResults_QUAC.MSAT[,1,], 1, mean)
v.com_means <- apply(samplingResults_QUAC.MSAT[,2,], 1, mean)
com_means <- apply(samplingResults_QUAC.MSAT[,3,], 1, mean)
lowfr_means <- apply(samplingResults_QUAC.MSAT[,4,], 1, mean)
rare_means <- apply(samplingResults_QUAC.MSAT[,5,], 1, mean)
# Pick plot colors (for all plots!), with transparency for values other than Total
plotColors <- c("red","red4","darkorange3","coral","purple")
plotColors[2:5] <- alpha(plotColors[2:5], 0.5)
# Plots all sets of points onto single graph, as well as 95% threshold line
plot(total_means, ylim=c(0,110), col=plotColors[1], pch=16, 
     xlab="Number of Individuals", ylab="Percent Diversity Capture",
     main="QUAC, Microsatellites (Complete): 50 Replicates")
points(v.com_means, col=plotColors[2], pch=16)
points(com_means, col=plotColors[3], pch=16)
points(lowfr_means, col=plotColors[4], pch=16)
points(rare_means, col=plotColors[5], pch=16)
legend(x=140, y=70.13276, inset = 0.05, legend = c("Total","Very common","Common","Low frequency", "Rare"),
       col=plotColors, pch = c(20,20,20), cex=1, pt.cex = 2, bty="n", y.intersp = 0.15)
# Lines for 95% threshold
abline(h=95, col="black", lty=3); abline(v=min_95_QUAC.MSAT, col="black")

# ---- SNPS (COMPLETE) ----
# READ IN GENIND FILE (QUAC DNFA; R0, min-maf=0; first SNP per locus; 2 populations, garden and wild) ----
genpop.filePath <- 
  "/RAID1/IMLS_GCCO/Analysis/Stacks/denovo_finalAssemblies/QUAC/output/populations_R0_NOMAF_1SNP_2Pops/"
setwd(genpop.filePath)
QUAC.SNP.genind <- read.genepop(paste0(genpop.filePath,"populations.snps.gen"))
# Correct popNames
pop(QUAC.SNP.genind) <- factor(read.table("QUAC_popmap_GardenWild", header=FALSE)[,2])
# Get QUAC SNP Tissue sample names, and rename SNP genind matrix
# File was created (by Austin K.), and can be found on Hoban Lab Drive ("MSATcomparisons_TissueNames")
QUAC.SNP.tissueNames_filepath <- paste0(SSRvSNP.wd,"exSituRepresentation/Resampling/QUAC_SNP_TissueNames.csv")
QUAC.SNP.tissueNames <- unlist(read.csv2(QUAC.SNP.tissueNames_filepath, header = TRUE, sep = ",")[3])
rownames(QUAC.SNP.genind@tab) <- QUAC.SNP.tissueNames

# CREATE SAMPLING RESULTS ARRAY ----
# Specify number of replicates
num_reps <- 50
# Export relevant functions and variables
clusterExport(cl, varlist = c("getAlleleCategories", "exSitu_Sample", "exSitu_Resample", 
                              "num_reps", "QUAC.SNP.genind"))
# Run resampling in parallel, to generate an array
samplingResults_QUAC.SNP <- 
  parSapply(cl, 1:num_reps, 
            function(a) exSitu_Resample(gen.obj = QUAC.SNP.genind), simplify = "array")
# Close cores
# stopCluster(cl)

# Average 95% minimum sampling size across replicates, as well as standard deviation
min_95_QUAC.SNP <- min(which(apply(samplingResults_QUAC.SNP[,1,],1,mean) > 95))
min_95.SD_QUAC.SNP <- apply(samplingResults_QUAC.SNP[,1,],1,sd)[min_95_QUAC.SNP]
print(min_95_QUAC.SNP); print(min_95.SD_QUAC.SNP)

# PLOTTING ----
# Calculate means and standard deviations, for each capture rate category
total_means <- apply(samplingResults_QUAC.SNP[,1,], 1, mean)
v.com_means <- apply(samplingResults_QUAC.SNP[,2,], 1, mean)
com_means <- apply(samplingResults_QUAC.SNP[,3,], 1, mean)
lowfr_means <- apply(samplingResults_QUAC.SNP[,4,], 1, mean)
rare_sd <- apply(samplingResults_QUAC.SNP[,5,], 1, sd)
# Plots all sets of points onto single graph, as well as 95% threshold line
plot(total_means, ylim=c(0,110), col=plotColors[1], pch=16, 
     xlab="Number of Individuals", ylab="Percent Diversity Capture",
     main="QUAC, SNPs (Complete): 50 Replicates")
points(v.com_means, col=plotColors[2], pch=16)
points(com_means, col=plotColors[3], pch=16)
points(lowfr_means, col=plotColors[4], pch=16)
points(rare_means, col=plotColors[5], pch=16)
legend(x=83, y=70.13276, inset = 0.05, legend = c("Total","Very common","Common","Low frequency", "Rare"),
       col=plotColors, pch = c(20,20,20), cex=1, pt.cex = 2, bty="n", y.intersp = 0.15)
# Lines for 95% threshold
abline(h=95, col="black", lty=3); abline(v=min_95_QUAC.SNP, col="black")

# ---- MSATS AND SNPS: SUBSET ----
# Subset SNP sample names by those that are also seen within the MSAT samples 
QUAC_sharedSamples <- sort(QUAC.SNP.tissueNames[which(QUAC.SNP.tissueNames %in% QUAC.MSAT.tissueNames)])
# Subset MSAT and SNP genind matrices to strictly shared samples, dropping now absent alleles
QUAC.MSAT_subset.genind <- QUAC.MSAT.genind[QUAC_sharedSamples,, drop=TRUE]
QUAC.SNP_subset.genind <- QUAC.SNP.genind[QUAC_sharedSamples,, drop=TRUE]

# CREATE SAMPLING RESULTS ARRAYS ----
# Specify number of replicates
num_reps <- 50
# Export relevant functions and variables
clusterExport(cl, varlist = c("getAlleleCategories", "exSitu_Sample", "exSitu_Resample", 
                              "num_reps", "QUAC.MSAT_subset.genind", "QUAC.SNP_subset.genind"))
# Run resampling in parallel, to generate arrays
samplingResults_QUAC.MSAT_subset <- 
  parSapply(cl, 1:num_reps, 
            function(a) exSitu_Resample(gen.obj = QUAC.MSAT_subset.genind), simplify = "array")

samplingResults_QUAC.SNP_subset <- 
  parSapply(cl, 1:num_reps, 
            function(a) exSitu_Resample(gen.obj = QUAC.SNP_subset.genind), simplify = "array")
# Close cores
# stopCluster(cl)

# MSAT_subset: average 95% minimum sampling size across replicates, as well as standard deviation
min_95_QUAC.MSAT_subset <- min(which(apply(samplingResults_QUAC.MSAT_subset[,1,],1,mean) > 95))
min_95.SD_QUAC.MSAT_subset <- apply(samplingResults_QUAC.MSAT_subset[,1,],1,sd)[min_95_QUAC.MSAT_subset]
print(min_95_QUAC.MSAT_subset); print(min_95.SD_QUAC.MSAT_subset)
# SNP_subset: average 95% minimum sampling size across replicates, as well as standard deviation
min_95_QUAC.SNP_subset <- min(which(apply(samplingResults_QUAC.SNP_subset[,1,],1,mean) > 95))
min_95.SD_QUAC.SNP_subset <- apply(samplingResults_QUAC.SNP_subset[,1,],1,sd)[min_95_QUAC.SNP_subset]
print(min_95_QUAC.SNP_subset); print(min_95.SD_QUAC.SNP_subset)

# PLOTTING ----
# TWO GRAPHS
# MSAT_subset: calculate means and standard deviations, for each capture rate category
total_means <- apply(samplingResults_QUAC.MSAT_subset[,1,], 1, mean)
v.com_means <- apply(samplingResults_QUAC.MSAT_subset[,2,], 1, mean)
com_means <- apply(samplingResults_QUAC.MSAT_subset[,3,], 1, mean)
lowfr_means <- apply(samplingResults_QUAC.MSAT_subset[,4,], 1, mean)
rare_sd <- apply(samplingResults_QUAC.MSAT_subset[,5,], 1, sd)
# Plots all sets of points onto single graph, as well as 95% threshold line
plot(total_means, ylim=c(0,110), col=plotColors[1], pch=16, 
     xlab="Number of Individuals", ylab="Percent Diversity Capture",
     main="QUAC, MSATs (Subset): 50 Replicates")
points(v.com_means, col=plotColors[2], pch=16)
points(com_means, col=plotColors[3], pch=16)
points(lowfr_means, col=plotColors[4], pch=16)
points(rare_means, col=plotColors[5], pch=16)
legend(x=80, y=70.13276, inset = 0.05, legend = c("Total","Very common","Common","Low frequency", "Rare"),
       col=plotColors, pch = c(20,20,20), cex=1, pt.cex = 2, bty="n", y.intersp = 0.15)
# Lines for 95% threshold
abline(h=95, col="black", lty=3); abline(v=min_95_QUAC.MSAT_subset, col="black")
# SNP_subset: calculate means and standard deviations, for each capture rate category
total_means <- apply(samplingResults_QUAC.SNP_subset[,1,], 1, mean)
v.com_means <- apply(samplingResults_QUAC.SNP_subset[,2,], 1, mean)
com_means <- apply(samplingResults_QUAC.SNP_subset[,3,], 1, mean)
lowfr_means <- apply(samplingResults_QUAC.SNP_subset[,4,], 1, mean)
rare_sd <- apply(samplingResults_QUAC.SNP_subset[,5,], 1, sd)
# Plots all sets of points onto single graph, as well as 95% threshold line
plot(total_means, ylim=c(0,110), col=plotColors[1], pch=16, 
     xlab="Number of Individuals", ylab="Percent Diversity Capture",
     main="QUAC, SNPs (Subset): 50 Replicates")
points(v.com_means, col=plotColors[2], pch=16)
points(com_means, col=plotColors[3], pch=16)
points(lowfr_means, col=plotColors[4], pch=16)
points(rare_means, col=plotColors[5], pch=16)
legend(x=83, y=70.13276, inset = 0.05, legend = c("Total","Very common","Common","Low frequency", "Rare"),
       col=plotColors, pch = c(20,20,20), cex=1, pt.cex = 2, bty="n", y.intersp = 0.15)
# Lines for 95% threshold
abline(h=95, col="black", lty=3); abline(v=min_95_QUAC.SNP_subset, col="black")

# ONE GRAPH (TOTAL VALUES)
# Relevant values
QUAC_total_means_MSAT_subset <- apply(samplingResults_QUAC.MSAT_subset[,1,], 1, mean)
min_95_QUAC.MSAT_subset <- min(which(apply(samplingResults_QUAC.MSAT_subset[,1,],1,mean) > 95))
QUAC_total_means_SNP_subset <- apply(samplingResults_QUAC.SNP_subset[,1,], 1, mean)
min_95_QUAC.SNP_subset <- min(which(apply(samplingResults_QUAC.SNP_subset[,1,],1,mean) > 95))
# Generate plot and add points
plotColors <- plotColors <- alpha(c("purple","darkgreen"), 0.7)
plot(QUAC_total_means_MSAT_subset, ylim=c(0,110), col=plotColors[1], pch=16, 
     xlab="Number of Individuals", ylab="Percent Diversity Capture",
     main="QUAC, MSATs and SNPs (Subset, 50 Replicates): Total Allelic Representation")
points(QUAC_total_means_SNP_subset, col=plotColors[2], pch=17)
# Legend
legend(x=81, y=70.13276, inset = 0.05, legend = c("Microsatellites","SNPs"),
       col=plotColors, pch = c(16,17), cex=1, pt.cex = 2, bty="n", y.intersp = 0.7)
# Lines for 95% threshold
abline(h=95, col="black", lty=3); abline(v=min_95_QUAC.MSAT_subset, col="black"); abline(v=min_95_QUAC.SNP_subset, col="black")

# %%%% QUBO %%%% ----
# ---- MSATS (COMPLETE) ----
# Read in genind file (Southeast Oaks repo; genetic_data/Qb_total.gen file)
genpop.filePath <- 
  "~/Documents/peripheralProjects/SE_oaks_genetics/genetic_data/"
setwd(genpop.filePath)
QUBO.MSAT.genind <- read.genepop(paste0(genpop.filePath,"Qb_total.gen"), ncode=3)
# Correct popNames: last population (IMLS4_MP1_IMLS336_C05) is garden; rest (9) are wild
levels(QUBO.MSAT.genind@pop) <- c(rep("wild",9), "garden") 
# Rename samples: Split sample names on underscore, and return 3rd element. This is the "IMLS" portion
QUBO.MSAT.sampleNames <- unlist(lapply(rownames(QUBO.MSAT.genind@tab), 
                                       function(x) strsplit(x, "_")[[1]][3]))
rownames(QUBO.MSAT.genind@tab) <- QUBO.MSAT.sampleNames

# CREATE SAMPLING RESULTS ARRAY ----
# Specify number of replicates
num_reps <- 50
# Export relevant functions and variables
clusterExport(cl, varlist = c("getAlleleCategories", "exSitu_Sample", "exSitu_Resample", 
                              "num_reps", "QUBO.MSAT.genind"))
# Run resampling in parallel, to generate an array
samplingResults_QUBO.MSAT <- 
  parSapply(cl, 1:num_reps, function(a) exSitu_Resample(gen.obj = QUBO.MSAT.genind), simplify = "array")

# Close cores
# stopCluster(cl)

# Average results across replicates (slices) of the sampling array, to determine
# the minimum number of samples required to capture 95% wild genetic diversity
# (We average samplingResults[,1,], since this column contains the total genetic diversity)
# Standard deviation is calculated as well
min_95_QUBO.MSAT <- min(which(apply(samplingResults_QUBO.MSAT[,1,],1,mean) > 95))
min_95.SD_QUBO.MSAT <- apply(samplingResults_QUBO.MSAT[,1,],1,sd)[min_95_QUBO.MSAT]
print(min_95_QUBO.MSAT); print(min_95.SD_QUBO.MSAT)

# PLOTTING ----
# Calculate means and standard deviations, for each capture rate category
total_means <- apply(samplingResults_QUBO.MSAT[,1,], 1, mean)
v.com_means <- apply(samplingResults_QUBO.MSAT[,2,], 1, mean)
com_means <- apply(samplingResults_QUBO.MSAT[,3,], 1, mean)
lowfr_means <- apply(samplingResults_QUBO.MSAT[,4,], 1, mean)
rare_means <- apply(samplingResults_QUBO.MSAT[,5,], 1, mean)
# Plots all sets of points onto single graph, as well as 95% threshold line
plot(total_means, ylim=c(0,110), col=plotColors[1], pch=16, 
     xlab="Number of Individuals", ylab="Percent Diversity Capture",
     main="QUBO, Microsatellites (Complete): 50 Replicates")
points(v.com_means, col=plotColors[2], pch=16)
points(com_means, col=plotColors[3], pch=16)
points(lowfr_means, col=plotColors[4], pch=16)
points(rare_means, col=plotColors[5], pch=16)
legend(x=200, y=70.13276, inset = 0.05, legend = c("Total","Very common","Common","Low frequency", "Rare"),
       col=plotColors, pch = c(20,20,20), cex=1, pt.cex = 2, bty="n", y.intersp = 0.15)
# Lines for 95% threshold
abline(h=95, col="black", lty=3); abline(v=min_95_QUBO.MSAT, col="black")

# ---- SNPS (COMPLETE) ----
# READ IN GENIND FILE (QUBO DNFA; R0, min-maf=0; first SNP per locus; 2 populations, garden and wild) ----
genpop.filePath <- 
  "/RAID1/IMLS_GCCO/Analysis/Stacks/reference_filteredReads/QUBO/GSNAP4/output/populations_R0_NOMAF_1SNP_2Pops/"
setwd(genpop.filePath)
QUBO.SNP.genind <- read.genepop(paste0(genpop.filePath,"populations.snps.gen"))
# Correct popNames
pop(QUBO.SNP.genind) <- factor(read.table("QUBO_popmap_GardenWild", header=FALSE)[,2])
# Rename samples: get rid of "QUBO_G/W_" prefix. This will just leave the "IMLS" portion
QUBO.SNP.sampleNames <- gsub("QUBO_G_",replacement = "", row.names(QUBO.SNP.genind@tab))
QUBO.SNP.sampleNames <- gsub("QUBO_W_",replacement = "", QUBO.SNP.sampleNames)
# Replace SH-Q names in SNP list with IMLS names
# These were determined by Austin K., and are outlined on the Hoban Lab Drive ("MSATcomparisons_TissueNames")
# Only 1 of the 11 SH_Q garden samples has an IMLS sample name (SHQ2177); others are unshared 
QUBO.SNP.sampleNames <- gsub("SH_Q2177",replacement = "IMLS338", QUBO.SNP.sampleNames)
QUBO.SNP.sampleNames <- gsub("SH_Q2178",replacement = "IMLS312", QUBO.SNP.sampleNames)
QUBO.SNP.sampleNames <- gsub("SH_Q2179",replacement = "IMLS062", QUBO.SNP.sampleNames)
QUBO.SNP.sampleNames <- gsub("SH_Q2180",replacement = "IMLS051", QUBO.SNP.sampleNames)
QUBO.SNP.sampleNames <- gsub("SH_Q2181",replacement = "IMLS011", QUBO.SNP.sampleNames)
QUBO.SNP.sampleNames <- gsub("SH_Q2182",replacement = "IMLS144", QUBO.SNP.sampleNames)
QUBO.SNP.sampleNames <- gsub("SH_Q2183",replacement = "IMLS170", QUBO.SNP.sampleNames)
QUBO.SNP.sampleNames <- gsub("SH_Q2184",replacement = "IMLS005", QUBO.SNP.sampleNames)
QUBO.SNP.sampleNames <- gsub("SH_Q2186",replacement = "IMLS017", QUBO.SNP.sampleNames)
# Rename sample matrix
rownames(QUBO.SNP.genind@tab) <- QUBO.SNP.sampleNames

# CREATE SAMPLING RESULTS ARRAY ----
# Specify number of replicates
num_reps <- 50
# Export relevant functions and variables
clusterExport(cl, varlist = c("getAlleleCategories", "exSitu_Sample", "exSitu_Resample", 
                              "num_reps", "QUBO.SNP.genind"))
# Run resampling in parallel, to generate an array
samplingResults_QUBO.SNP <- 
  parSapply(cl, 1:num_reps, function(a) exSitu_Resample(gen.obj = QUBO.SNP.genind), simplify = "array")
# Close cores
# stopCluster(cl)

# Average results across replicates (slices) of the sampling array, to determine
# the minimum number of samples required to capture 95% wild genetic diversity
# (We average samplingResults[,1,], since this column contains the total genetic diversity)
# Standard deviation is calculated as well
min_95_QUBO.SNP <- min(which(apply(samplingResults_QUBO.SNP[,1,],1,mean) > 95))
min_95.SD_QUBO.SNP <- apply(samplingResults_QUBO.SNP[,1,],1,sd)[min_95_QUBO.SNP]
print(min_95_QUBO.SNP); print(min_95.SD_QUBO.SNP)

# PLOTTING ----
# Calculate means and standard deviations, for each capture rate category
total_means <- apply(samplingResults_QUBO.SNP[,1,], 1, mean)
v.com_means <- apply(samplingResults_QUBO.SNP[,2,], 1, mean)
com_means <- apply(samplingResults_QUBO.SNP[,3,], 1, mean)
lowfr_means <- apply(samplingResults_QUBO.SNP[,4,], 1, mean)
rare_means <- apply(samplingResults_QUBO.SNP[,5,], 1, mean)
# Plots all sets of points onto single graph, as well as 95% threshold line
plot(total_means, ylim=c(0,110), col=plotColors[1], pch=16, 
     xlab="Number of Individuals", ylab="Percent Diversity Capture",
     main="QUBO, Microsatellites (Complete): 50 Replicates")
points(v.com_means, col=plotColors[2], pch=16)
points(com_means, col=plotColors[3], pch=16)
points(lowfr_means, col=plotColors[4], pch=16)
points(rare_means, col=plotColors[5], pch=16)
legend(x=83, y=70.13276, inset = 0.05, legend = c("Total","Very common","Common","Low frequency", "Rare"),
       col=plotColors, pch = c(20,20,20), cex=1, pt.cex = 2, bty="n", y.intersp = 0.15)
# Lines for 95% threshold
abline(h=95, col="black", lty=3); abline(v=min_95_QUBO.SNP, col="black")

# ---- MSATS AND SNPS: SUBSET ----
# Subset SNP sample names by those that are also seen within the MSAT samples
QUBO_sharedSamples <- sort(QUBO.SNP.sampleNames[which(QUBO.SNP.sampleNames %in% QUBO.MSAT.sampleNames)])
# Subset MSAT and SNP wild matrix objects to strictly shared samples
QUBO.MSAT_subset.genind <- QUBO.MSAT.genind[QUBO_sharedSamples,, drop=TRUE]
QUBO.SNP_subset.genind <- QUBO.SNP.genind[QUBO_sharedSamples,, drop=TRUE]

# Specify number of replicates
num_reps <- 50
# Export relevant functions and variables
clusterExport(cl, varlist = c("getAlleleCategories", "exSitu_Sample", "exSitu_Resample", 
                              "num_reps", "QUBO.MSAT_subset.genind", "QUBO.SNP_subset.genind"))
# Run resampling in parallel, to generate arrays
samplingResults_QUBO.MSAT_subset <- 
  parSapply(cl, 1:num_reps, 
            function(a) exSitu_Resample(gen.obj = QUBO.MSAT_subset.genind), simplify = "array")

samplingResults_QUBO.SNP_subset <- 
  parSapply(cl, 1:num_reps, 
            function(a) exSitu_Resample(gen.obj = QUBO.SNP_subset.genind), simplify = "array")
# Close cores
stopCluster(cl)

# MSAT_subset: average 95% minimum sampling size across replicates, as well as standard deviation
min_95_QUBO.MSAT_subset <- min(which(apply(samplingResults_QUBO.MSAT_subset[,1,],1,mean) > 95))
min_95.SD_QUBO.MSAT_subset <- apply(samplingResults_QUBO.MSAT_subset[,1,],1,sd)[min_95_QUBO.MSAT_subset]
print(min_95_QUBO.MSAT_subset); print(min_95.SD_QUBO.MSAT_subset)
# SNP_subset: average 95% minimum sampling size across replicates, as well as standard deviation
min_95_QUBO.SNP_subset <- min(which(apply(samplingResults_QUBO.SNP_subset[,1,],1,mean) > 95))
min_95.SD_QUBO.SNP_subset <- apply(samplingResults_QUBO.SNP_subset[,1,],1,sd)[min_95_QUBO.SNP_subset]
print(min_95_QUBO.SNP_subset); print(min_95.SD_QUBO.SNP_subset)

# PLOTTING ----
# TWO GRAPHS
# MSAT_subset: calculate means and standard deviations, for each capture rate category
total_means <- apply(samplingResults_QUBO.MSAT_subset[,1,], 1, mean)
v.com_means <- apply(samplingResults_QUBO.MSAT_subset[,2,], 1, mean)
com_means <- apply(samplingResults_QUBO.MSAT_subset[,3,], 1, mean)
lowfr_means <- apply(samplingResults_QUBO.MSAT_subset[,4,], 1, mean)
rare_sd <- apply(samplingResults_QUBO.MSAT_subset[,5,], 1, sd)
# Plots all sets of points onto single graph, as well as 95% threshold line
plot(total_means, ylim=c(0,110), col=plotColors[1], pch=16, 
     xlab="Number of Individuals", ylab="Percent Diversity Capture",
     main="QUBO, MSATs (Subset): 50 Replicates")
points(v.com_means, col=plotColors[2], pch=16)
points(com_means, col=plotColors[3], pch=16)
points(lowfr_means, col=plotColors[4], pch=16)
points(rare_means, col=plotColors[5], pch=16)
legend(x=80, y=70.13276, inset = 0.05, legend = c("Total","Very common","Common","Low frequency", "Rare"),
       col=plotColors, pch = c(20,20,20), cex=1, pt.cex = 2, bty="n", y.intersp = 0.15)
# Lines for 95% threshold
abline(h=95, col="black", lty=3); abline(v=min_95_QUBO.MSAT_subset, col="black")
# SNP_subset: calculate means and standard deviations, for each capture rate category
total_means <- apply(samplingResults_QUBO.SNP_subset[,1,], 1, mean)
v.com_means <- apply(samplingResults_QUBO.SNP_subset[,2,], 1, mean)
com_means <- apply(samplingResults_QUBO.SNP_subset[,3,], 1, mean)
lowfr_means <- apply(samplingResults_QUBO.SNP_subset[,4,], 1, mean)
rare_sd <- apply(samplingResults_QUBO.SNP_subset[,5,], 1, sd)
# Pick colors, with transparency for values other than Total
plotColors <- c("red","red4","darkorange3","coral","purple")
plotColors[2:5] <- alpha(plotColors[2:5], 0.5)
# Plots all sets of points onto single graph, as well as 95% threshold line
plot(total_means, ylim=c(0,110), col=plotColors[1], pch=16, 
     xlab="Number of Individuals", ylab="Percent Diversity Capture",
     main="QUBO, SNPs (Subset): 50 Replicates")
points(v.com_means, col=plotColors[2], pch=16)
points(com_means, col=plotColors[3], pch=16)
points(lowfr_means, col=plotColors[4], pch=16)
points(rare_means, col=plotColors[5], pch=16)
legend(x=83, y=70.13276, inset = 0.05, legend = c("Total","Very common","Common","Low frequency", "Rare"),
       col=plotColors, pch = c(20,20,20), cex=1, pt.cex = 2, bty="n", y.intersp = 0.15)
# Lines for 95% threshold
abline(h=95, col="black", lty=3); abline(v=min_95_QUBO.SNP_subset, col="black")

# ONE GRAPH (TOTAL VALUES)
# Relevant values
QUBO_total_means_MSAT_subset <- apply(samplingResults_QUBO.MSAT_subset[,1,], 1, mean)
min_95_QUBO.MSAT_subset <- min(which(apply(samplingResults_QUBO.MSAT_subset[,1,],1,mean) > 95))
QUBO_total_means_SNP_subset <- apply(samplingResults_QUBO.SNP_subset[,1,], 1, mean)
min_95_QUBO.SNP_subset <- min(which(apply(samplingResults_QUBO.SNP_subset[,1,],1,mean) > 95))
# Generate plot and add points
plotColors <- plotColors <- alpha(c("purple","darkgreen"), 0.7)
plot(QUBO_total_means_MSAT_subset, ylim=c(0,110), col=plotColors[1], pch=16, 
     xlab="Number of Individuals", ylab="Percent Diversity Capture",
     main="QUBO, MSATs and SNPs (Subset, 50 Replicates): Total Allelic Representation")
points(QUBO_total_means_SNP_subset, col=plotColors[2], pch=17)
# Legend
legend(x=78, y=70.13276, inset = 0.05, legend = c("Microsatellites","SNPs"),
       col=plotColors, pch = c(16,17), cex=1, pt.cex = 2, bty="n", y.intersp = 0.7)
# Lines for 95% threshold
abline(h=95, col="black", lty=3); abline(v=min_95_QUBO.MSAT_subset, col="black"); abline(v=min_95_QUBO.SNP_subset, col="black")
